{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "745fda9f",
   "metadata": {},
   "source": [
    "# Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef174cfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2026-01-16 16:59:54 | [INFO] cffi mode is CFFI_MODE.ANY\n",
      "2026-01-16 16:59:54 | [INFO] R home found: /usr/lib/R\n",
      "2026-01-16 16:59:55 | [INFO] R library path: /usr/lib/R/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/default-java/lib/server:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64\n",
      "2026-01-16 16:59:55 | [INFO] LD_LIBRARY_PATH: /usr/lib/R/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/default-java/lib/server:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64\n",
      "2026-01-16 16:59:55 | [INFO] Default options to initialize R: rpy2, --quiet, --no-save\n",
      "2026-01-16 16:59:55 | [INFO] Environment variable \"LD_LIBRARY_PATH\" redefined by R and overriding existing variable. Current: \"/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64\", R: \"/usr/lib/R/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/default-java/lib/server:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64:/usr/local/cuda-12.2/lib64\"\n",
      "2026-01-16 16:59:55 | [INFO] Environment variable \"PWD\" redefined by R and overriding existing variable. Current: \"/home/gdallagl\", R: \"/home/gdallagl/myworkdir/XDP/script/multi_sample_analysis\"\n",
      "2026-01-16 16:59:55 | [INFO] R is already initialized. No need to initialize.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import decoupler as dc\n",
    "\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "from utils import preprocessing\n",
    "from utils import DEG\n",
    "\n",
    "\n",
    "BASE_PATH = os.getenv(\"BASE_PATH_FOR_SAVING\")\n",
    "DISEASED_SAMPLES_FOLDER = f\"{BASE_PATH}/diseased\"\n",
    "MSN_BICAN_FOLDER = \"/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad\"\n",
    "\n",
    "# Cell types to Analyse by GroupName in mmc\n",
    "CT_TO_ANALYSE = [\n",
    "\"STRd D1 Matrix MSN\",\"STRv D1 MSN\",\"Tafa1 D1 Matrix MSN\",\"STRd D2 Matrix MSN\",\"STRv D2 MSN\",\n",
    "\"Tafa1 D2 Matrix MSN\",\"STRd D1 Striosome MSN\",\"STRd D2 Striosome MSN\",\"STRv D1 NUDAP MSN\",\"STR D1D2 Hybrid MSN\",\n",
    "\"STRd D2 StrioMat Hybrid MSN\",\"Tafa1 STRd D2 StrioMat Hybrid MSN\"]#,\"OPC\",\"Astrocyte\"]\n",
    "\n",
    "# Cell type annotation to use for DEG\n",
    "CT_ANNOTATION_FOR_DEG = \"\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab42d2a7",
   "metadata": {},
   "source": [
    "# Create Healthy Diseased merged adata\n",
    "\n",
    "Create adata obj with both dieseased anf helathy.\n",
    "\n",
    "Helthy and diseased samples are saved in differt locations and have differ nomenclture for columns.\n",
    "\n",
    "In addition, diseased are split btu sample, while healthy are already split by cell type but not by sample."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43e5f5fb",
   "metadata": {},
   "source": [
    "### Merge Diseased adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6c8f22d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/data/XDP/diseased/sample_01\n",
      "/home/gdallagl/myworkdir/data/XDP/diseased/recon_241105\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[View of AnnData object with n_obs × n_vars = 11130 × 32780\n",
       "     obs: 'x', 'y', 'pct_intronic', 'is_cell', 'dbscan_clusters', 'dbscan_score', 'has_spatial', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Neighborhood_aggregate_probability', 'Class_name', 'Class_bootstrapping_probability', 'Class_aggregate_probability', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Subclass_aggregate_probability', 'Group_name', 'Group_bootstrapping_probability', 'Group_aggregate_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'my_hierarchy_all', 'library', 'library_name', 'leiden_1', 'leiden_2', 'leiden_3', 'leiden_4', 'leiden_scANVI', 'mapping_MSN_for_zoning', 'mapping_coupling_for_zoning', 'zone', 'zone_probability'\n",
       "     var: 'feature_types', 'genome', 'gene_symbol', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'highly_variable_nbatches', 'mean', 'std',\n",
       " View of AnnData object with n_obs × n_vars = 19714 × 36190\n",
       "     obs: 'total_counts', 'n_genes', 'logumi', 'pct_counts_mt', 'seurat_clusters', 'x', 'y', 'library', 'cell_class', 'cell_subclass', 'reclustered_patch_matrix_exotic', 'has_spatial', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Neighborhood_aggregate_probability', 'Neighborhood_correlation_coefficient', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Class_aggregate_probability', 'Class_correlation_coefficient', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Subclass_aggregate_probability', 'Subclass_correlation_coefficient', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Group_aggregate_probability', 'Group_correlation_coefficient', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'Cluster_aggregate_probability', 'Cluster_correlation_coefficient', 'pct_intronic', 'is_cell', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'leiden_1', 'leiden_2', 'leiden_3', 'leiden_4', 'leiden_5', 'cluster_entropy', 'is_mixed_cluster_doublet', 'Astrocyte_score', 'Ependymal_score', 'Endothelial_score', 'Lymphocyte_score', 'Microglia_score', 'Macrophage_score', 'Oligodendrocyte_score', 'OPC_score', 'Neuron_inhibitory_score', 'Neuron_excitatory_score', 'Neuron_dopaminergic_score', 'Neuron_cholinergic_score', 'Neuron_serotonergic_score', 'is_oligo_doublet', 'scrublet_score', 'is_scrublet_doublet', 'is_doublet', 'my_hierarchy_all', 'grouping_for_MSN_QCed', 'mapping_coupling_for_zoning', 'zone', 'zone_probability'\n",
       "     var: 'gene_symbol', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm', 'mean', 'std']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_diseased_list = []\n",
    "sample_diseased_list = []\n",
    "\n",
    "# Load Diseased data (each adata is a differt sample/patient)\n",
    "for sample_folder in [p for p in Path(DISEASED_SAMPLES_FOLDER).iterdir() if p.is_dir()]:\n",
    "\n",
    "    print(sample_folder)\n",
    "\n",
    "    adata_path = sample_folder / \"adata\" / \"all_lib_adata_zoned.h5ad\"\n",
    "\n",
    "    adata_tmp = sc.read_h5ad(adata_path, backed=\"r\")\n",
    "\n",
    "    # print(adata_tmp.obs.columns)\n",
    "    # display(adata_tmp.obs)\n",
    "    # print(sample_folder)\n",
    "\n",
    "    # restart from raw counts\n",
    "    adata_minimal = sc.AnnData(\n",
    "        X=adata_tmp.layers[\"counts\"].copy(),                   \n",
    "        obs=adata_tmp.obs.copy(),\n",
    "        var=adata_tmp.var.copy()\n",
    "    )\n",
    "    del adata_tmp\n",
    "\n",
    "    # Very that .X is raw counts\n",
    "    all_integer_like = np.all(np.mod(adata_minimal.X.data, 1) == 0)\n",
    "    assert all_integer_like, f\"Expression matrix .X is not integer-like (raw counts). --> {adata_path}\"\n",
    "\n",
    "    # Filter cell types\n",
    "    adata_minimal = adata_minimal[adata_minimal.obs[\"Group_name\"].isin(CT_TO_ANALYSE)]\n",
    "\n",
    "    # add sample:\n",
    "    sample_diseased_list.append(sample_folder.name)\n",
    "\n",
    "    # add ot list\n",
    "    adata_diseased_list.append(adata_minimal)\n",
    "\n",
    "    #break\n",
    "\n",
    "adata_diseased_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "46c3698f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 30844 × 30914\n",
      "    obs: 'x', 'y', 'pct_intronic', 'is_cell', 'has_spatial', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Neighborhood_aggregate_probability', 'Class_name', 'Class_bootstrapping_probability', 'Class_aggregate_probability', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Subclass_aggregate_probability', 'Group_name', 'Group_bootstrapping_probability', 'Group_aggregate_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'my_hierarchy_all', 'library', 'leiden_1', 'leiden_2', 'leiden_3', 'leiden_4', 'mapping_coupling_for_zoning', 'zone', 'zone_probability', 'donor_id'\n",
      "    var: 'feature_types', 'genome', 'gene_symbol'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:865: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    ['MIR1302-2HG', 'ENSG00000238009', 'ENSG00000239945', 'ENSG00000239906', 'ENSG00000241860']\n",
      "\n",
      "    Inferred to be: categorical\n",
      "\n",
      "  names = self._prep_dim_index(names, \"var\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>pct_intronic</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>has_spatial</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>Neighborhood_aggregate_probability</th>\n",
       "      <th>Class_name</th>\n",
       "      <th>Class_bootstrapping_probability</th>\n",
       "      <th>...</th>\n",
       "      <th>my_hierarchy_all</th>\n",
       "      <th>library</th>\n",
       "      <th>leiden_1</th>\n",
       "      <th>leiden_2</th>\n",
       "      <th>leiden_3</th>\n",
       "      <th>leiden_4</th>\n",
       "      <th>mapping_coupling_for_zoning</th>\n",
       "      <th>zone</th>\n",
       "      <th>zone_probability</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAACCAGCACTAGTCG-1-SI-TT-H1-sample_01</th>\n",
       "      <td>47925.552952</td>\n",
       "      <td>-3509.217099</td>\n",
       "      <td>0.647217</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>SPN</td>\n",
       "      <td>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>STR D1 Matrix MSN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.819</td>\n",
       "      <td>sample_01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAACCCCAGCCATAAT-1-SI-TT-H1-sample_01</th>\n",
       "      <td>39898.157546</td>\n",
       "      <td>394.065473</td>\n",
       "      <td>0.596793</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>SPN</td>\n",
       "      <td>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>25</td>\n",
       "      <td>STR D2 Matrix MSN</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>sample_01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAACTAGAGCCTAGAC-1-SI-TT-H1-sample_01</th>\n",
       "      <td>2380.507310</td>\n",
       "      <td>2218.759126</td>\n",
       "      <td>0.657474</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>SPN</td>\n",
       "      <td>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>sample_01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               x            y  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  47925.552952 -3509.217099   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  39898.157546   394.065473   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   2380.507310  2218.759126   \n",
       "\n",
       "                                                    pct_intronic  is_cell  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      0.647217     True   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      0.596793     True   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      0.657474     True   \n",
       "\n",
       "                                                    has_spatial  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...         True   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...         True   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...        False   \n",
       "\n",
       "                                                   Neighborhood_name  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   Subpallium GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   Subpallium GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   Subpallium GABA   \n",
       "\n",
       "                                                    Neighborhood_bootstrapping_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                     1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                     1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                     1.0   \n",
       "\n",
       "                                                    Neighborhood_aggregate_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                 1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                 1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                 1.0   \n",
       "\n",
       "                                                     Class_name  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CN LGE GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CN LGE GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CN LGE GABA   \n",
       "\n",
       "                                                    Class_bootstrapping_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "\n",
       "                                                    ...  my_hierarchy_all  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  ...               SPN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  ...               SPN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  ...               SPN   \n",
       "\n",
       "                                                                                     library  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1   \n",
       "\n",
       "                                                    leiden_1  leiden_2  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...         4         7   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...         6         8   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...         7        14   \n",
       "\n",
       "                                                   leiden_3  leiden_4  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...       13        20   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...       16        25   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...       24        33   \n",
       "\n",
       "                                                    mapping_coupling_for_zoning  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...            STR D1 Matrix MSN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...            STR D2 Matrix MSN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...            STRv D1 NUDAP MSN   \n",
       "\n",
       "                                                   zone zone_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...    3            0.819   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...    5            1.000   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...    3            1.000   \n",
       "\n",
       "                                                     donor_id  \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  sample_01  \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  sample_01  \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  sample_01  \n",
       "\n",
       "[3 rows x 38 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   feature_types  genome      gene_symbol\n",
       "gene_symbol                                              \n",
       "MIR1302-2HG      Gene Expression  GRCh38      MIR1302-2HG\n",
       "ENSG00000238009  Gene Expression  GRCh38  ENSG00000238009\n",
       "ENSG00000239945  Gene Expression  GRCh38  ENSG00000239945"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Save var info from first dataset before concat\n",
    "# So all dataset will have same metadata for genes\n",
    "var_info = adata_diseased_list[0].var[['feature_types', 'genome', 'gene_symbol']].copy()\n",
    "\n",
    "# Concate\n",
    "all_adata_diseased = sc.concat(adata_diseased_list, \n",
    "                  axis=0, # concatenate along observations, i.e., stacking cells)\n",
    "                  index_unique='-' , # causes new barcodes to be orig_idx + '-' + key\n",
    "                  keys=sample_diseased_list,\n",
    "                  label=\"donor_id\",\n",
    "                  join='inner')  # keeps COMMON genes\n",
    "del adata_diseased_list\n",
    "\n",
    "# Restore var info for common genes\n",
    "all_adata_diseased.var = var_info.loc[all_adata_diseased.var_names]\n",
    "\n",
    "# Gene naems need to be symbols to be the same as healthy adata\n",
    "all_adata_diseased.var_names = all_adata_diseased.var['gene_symbol']\n",
    "\n",
    "print(all_adata_diseased)\n",
    "display(all_adata_diseased.obs.head(3))\n",
    "display(all_adata_diseased.var.head(3))\n",
    "assert all_adata_diseased.obs_names.is_unique, \"ERROR: all_adata_diseased.obs_names are not unique!\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44edae10",
   "metadata": {},
   "source": [
    "### Merge Healthy adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "793f9bd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STRd_D2_Striosome_MSN.h5ad\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STRd_D1_Striosome_MSN.h5ad\n",
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STRd_D2_StrioMat_Hybrid_MSN.h5ad\n",
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/Astrocyte.h5ad\n",
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STRv_D1_NUDAP_MSN.h5ad\n",
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STR_D1_Matrix_MSN.h5ad\n",
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STR_D2_Matrix_MSN.h5ad\n",
      "/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STR_D1D2_Hybrid_MSN.h5ad\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[View of AnnData object with n_obs × n_vars = 17130 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'source_file', 'special_celltype', 'GM', 'ident', 'D1_zones_6_assigned_v2',\n",
       " View of AnnData object with n_obs × n_vars = 20903 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'source_file', 'hc', 'GM', 'ident', 'D1_zones_6_assigned_v2',\n",
       " View of AnnData object with n_obs × n_vars = 4325 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'source_file', 'GM', 'ident', 'D1_zones_6_assigned_v2',\n",
       " View of AnnData object with n_obs × n_vars = 0 × 37902\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'x', 'y', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'x_um2', 'y_um2', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'x_um', 'y_um', 'harmony_clusters', 'RNA_snn_res.0.4', 'GM', 'put_cluster', 'D1_zones_6_assigned_v1', 'D1_zones_6_assigned_v2', 'D1_zones_6_assigned', 'put_cluster2', 'GM_Putamen_Cluster', 'ident',\n",
       " View of AnnData object with n_obs × n_vars = 3514 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'source_file', 'GM', 'ident', 'D1_zones_6_assigned_v2',\n",
       " View of AnnData object with n_obs × n_vars = 60598 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'logUMI', 'source_file', 'special_celltype', 'GM', 'ident', 'D1_zones_6_assigned_v2',\n",
       " View of AnnData object with n_obs × n_vars = 61035 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'source_file', 'special_celltype', 'GM', 'ident', 'D1_zones_6_assigned_v2',\n",
       " View of AnnData object with n_obs × n_vars = 12067 × 37905\n",
       "     obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'source_file', 'GM', 'ident', 'D1_zones_6_assigned_v2']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_healthy_list = []\n",
    "\n",
    "# Load Helahy data (each adata is a differt cell type)\n",
    "for adata_path in [p for p in Path(MSN_BICAN_FOLDER).iterdir() if not p.is_dir()]:\n",
    "\n",
    "    print(adata_path)\n",
    "\n",
    "    adata_tmp = sc.read_h5ad(adata_path)#, backed=\"r\")\n",
    "\n",
    "    # print(adata_tmp.obs.columns)\n",
    "    # display(adata_tmp.obs)\n",
    "    # print(sample_folder)\n",
    "\n",
    "    # restart from raw counts\n",
    "    adata_minimal = sc.AnnData(\n",
    "        X=adata_tmp.X.copy(),           # raw counts in X       \n",
    "        obs=adata_tmp.obs.copy(),\n",
    "        var=adata_tmp.var.copy()\n",
    "    )\n",
    "    del adata_tmp\n",
    "\n",
    "        # Filter cell types\n",
    "    adata_minimal = adata_minimal[adata_minimal.obs[\"Group_name\"].isin(CT_TO_ANALYSE)]\n",
    "\n",
    "    # Very that .X is raw counts\n",
    "    all_integer_like = np.all(np.mod(adata_minimal.X.data, 1) == 0)\n",
    "    assert all_integer_like, f\"Expression matrix .X is not integer-like (raw counts). --> {adata_path}\"\n",
    "\n",
    "    # add ot list\n",
    "    adata_healthy_list.append(adata_minimal)\n",
    "\n",
    "adata_healthy_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "089b881c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 179572 × 37902\n",
      "    obs: 'orig.ident', 'nCount_originalexp', 'nFeature_originalexp', 'PREFIX', 'CELL_BARCODE', 'NUM_GENIC_READS', 'NUM_TRANSCRIPTS', 'NUM_GENES', 'num_retained_transcripts', 'pct_coding', 'pct_utr', 'pct_intergenic', 'pct_genic', 'pct_intronic', 'pct_mt', 'pct_ribosomal', 'frac_contamination', 'experiment', 'donor_external_id', 'is_primary_data', 'Neighborhood_label', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_label', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_label', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_label', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'donor_id', 'n_counts', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.8', 'seurat_clusters', 'library', 'donor', 'unique_cell_ID', 'AK_celltype', 'RNA_snn_res.0.2', 'cb', 'clusters', 'num_dbscan', 'logumi', 'cb_index', 'percent.mt', 'harmony_clusters', 'GM', 'ident', 'D1_zones_6_assigned_v2', 'zone', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt'\n",
      "    var: 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_originalexp</th>\n",
       "      <th>nFeature_originalexp</th>\n",
       "      <th>PREFIX</th>\n",
       "      <th>CELL_BARCODE</th>\n",
       "      <th>NUM_GENIC_READS</th>\n",
       "      <th>NUM_TRANSCRIPTS</th>\n",
       "      <th>NUM_GENES</th>\n",
       "      <th>num_retained_transcripts</th>\n",
       "      <th>pct_coding</th>\n",
       "      <th>...</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>log1p_total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>s5_s5_CGTTGGTAGATCCAAT_1-0</th>\n",
       "      <td>2024-10-25</td>\n",
       "      <td>144267.0</td>\n",
       "      <td>11179</td>\n",
       "      <td>2024-10-25_s5_Slide-tag_10X-GEMX-5P-GEX_BN_rxn1</td>\n",
       "      <td>CGTTGGTAGATCCAAT</td>\n",
       "      <td>262799.0</td>\n",
       "      <td>139798.0</td>\n",
       "      <td>10686.0</td>\n",
       "      <td>139539.0</td>\n",
       "      <td>0.1612</td>\n",
       "      <td>...</td>\n",
       "      <td>9.321882</td>\n",
       "      <td>144267.0</td>\n",
       "      <td>11.879428</td>\n",
       "      <td>24.225221</td>\n",
       "      <td>31.915823</td>\n",
       "      <td>40.788261</td>\n",
       "      <td>54.600151</td>\n",
       "      <td>60.0</td>\n",
       "      <td>4.110874</td>\n",
       "      <td>0.041590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s5_s5_AACGCACTCCCCTATG_2-0</th>\n",
       "      <td>2024-10-25</td>\n",
       "      <td>33064.0</td>\n",
       "      <td>7429</td>\n",
       "      <td>2024-10-25_s5_Slide-tag_10X-GEMX-5P-GEX_BN_rxn2</td>\n",
       "      <td>AACGCACTCCCCTATG</td>\n",
       "      <td>68618.0</td>\n",
       "      <td>32512.0</td>\n",
       "      <td>7260.0</td>\n",
       "      <td>32112.0</td>\n",
       "      <td>0.2329</td>\n",
       "      <td>...</td>\n",
       "      <td>8.913281</td>\n",
       "      <td>33064.0</td>\n",
       "      <td>10.406231</td>\n",
       "      <td>21.691265</td>\n",
       "      <td>28.097024</td>\n",
       "      <td>35.930317</td>\n",
       "      <td>49.283208</td>\n",
       "      <td>28.0</td>\n",
       "      <td>3.367296</td>\n",
       "      <td>0.084684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s5_s5_CTAAGCGGTCCGAAAT_2-0</th>\n",
       "      <td>2024-10-25</td>\n",
       "      <td>350285.0</td>\n",
       "      <td>14106</td>\n",
       "      <td>2024-10-25_s5_Slide-tag_10X-GEMX-5P-GEX_BN_rxn2</td>\n",
       "      <td>CTAAGCGGTCCGAAAT</td>\n",
       "      <td>694067.0</td>\n",
       "      <td>335038.0</td>\n",
       "      <td>13281.0</td>\n",
       "      <td>334810.0</td>\n",
       "      <td>0.1425</td>\n",
       "      <td>...</td>\n",
       "      <td>9.554426</td>\n",
       "      <td>350285.0</td>\n",
       "      <td>12.766505</td>\n",
       "      <td>21.112237</td>\n",
       "      <td>28.276404</td>\n",
       "      <td>37.604522</td>\n",
       "      <td>52.113565</td>\n",
       "      <td>319.0</td>\n",
       "      <td>5.768321</td>\n",
       "      <td>0.091069</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            orig.ident  nCount_originalexp  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0  2024-10-25            144267.0   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0  2024-10-25             33064.0   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0  2024-10-25            350285.0   \n",
       "\n",
       "                            nFeature_originalexp  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0                 11179   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0                  7429   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0                 14106   \n",
       "\n",
       "                                                                     PREFIX  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0  2024-10-25_s5_Slide-tag_10X-GEMX-5P-GEX_BN_rxn1   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0  2024-10-25_s5_Slide-tag_10X-GEMX-5P-GEX_BN_rxn2   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0  2024-10-25_s5_Slide-tag_10X-GEMX-5P-GEX_BN_rxn2   \n",
       "\n",
       "                                CELL_BARCODE  NUM_GENIC_READS  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0  CGTTGGTAGATCCAAT         262799.0   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0  AACGCACTCCCCTATG          68618.0   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0  CTAAGCGGTCCGAAAT         694067.0   \n",
       "\n",
       "                            NUM_TRANSCRIPTS  NUM_GENES  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0         139798.0    10686.0   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0          32512.0     7260.0   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0         335038.0    13281.0   \n",
       "\n",
       "                            num_retained_transcripts  pct_coding  ...  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0                  139539.0      0.1612  ...   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0                   32112.0      0.2329  ...   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0                  334810.0      0.1425  ...   \n",
       "\n",
       "                            log1p_n_genes_by_counts  total_counts  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0                 9.321882      144267.0   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0                 8.913281       33064.0   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0                 9.554426      350285.0   \n",
       "\n",
       "                            log1p_total_counts  pct_counts_in_top_50_genes  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0           11.879428                   24.225221   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0           10.406231                   21.691265   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0           12.766505                   21.112237   \n",
       "\n",
       "                            pct_counts_in_top_100_genes  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0                    31.915823   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0                    28.097024   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0                    28.276404   \n",
       "\n",
       "                            pct_counts_in_top_200_genes  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0                    40.788261   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0                    35.930317   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0                    37.604522   \n",
       "\n",
       "                            pct_counts_in_top_500_genes total_counts_mt  \\\n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0                    54.600151            60.0   \n",
       "s5_s5_AACGCACTCCCCTATG_2-0                    49.283208            28.0   \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0                    52.113565           319.0   \n",
       "\n",
       "                           log1p_total_counts_mt  pct_counts_mt  \n",
       "s5_s5_CGTTGGTAGATCCAAT_1-0              4.110874       0.041590  \n",
       "s5_s5_AACGCACTCCCCTATG_2-0              3.367296       0.084684  \n",
       "s5_s5_CTAAGCGGTCCGAAAT_2-0              5.768321       0.091069  \n",
       "\n",
       "[3 rows x 69 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>False</td>\n",
       "      <td>33370</td>\n",
       "      <td>0.227496</td>\n",
       "      <td>0.204977</td>\n",
       "      <td>81.416925</td>\n",
       "      <td>40852.0</td>\n",
       "      <td>10.617736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>False</td>\n",
       "      <td>78478</td>\n",
       "      <td>0.689695</td>\n",
       "      <td>0.524548</td>\n",
       "      <td>56.297196</td>\n",
       "      <td>123850.0</td>\n",
       "      <td>11.726835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>False</td>\n",
       "      <td>2836</td>\n",
       "      <td>0.017436</td>\n",
       "      <td>0.017286</td>\n",
       "      <td>98.420689</td>\n",
       "      <td>3131.0</td>\n",
       "      <td>8.049427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mt  n_cells_by_counts  mean_counts  log1p_mean_counts  \\\n",
       "A1BG      False              33370     0.227496           0.204977   \n",
       "A1BG-AS1  False              78478     0.689695           0.524548   \n",
       "A1CF      False               2836     0.017436           0.017286   \n",
       "\n",
       "          pct_dropout_by_counts  total_counts  log1p_total_counts  \n",
       "A1BG                  81.416925       40852.0           10.617736  \n",
       "A1BG-AS1              56.297196      123850.0           11.726835  \n",
       "A1CF                  98.420689        3131.0            8.049427  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Concate\n",
    "all_adata_healthy = sc.concat(adata_healthy_list, \n",
    "                  axis=0, # concatenate along observations, i.e., stacking cells)\n",
    "                  index_unique='-' , # causes new barcodes to be orig_idx + '-' + key\n",
    "                #   keys=sample_diseased_list,\n",
    "                #   label=\"donor_id\",\n",
    "                  join='inner')  # keeps COMMON genes\n",
    "del adata_healthy_list\n",
    "\n",
    "# Rename columns to match diseased\n",
    "all_adata_healthy.obs['zone'] = all_adata_healthy.obs['harmony_clusters'].astype(str).copy()\n",
    "\n",
    "# claculte pythonic QC\n",
    "all_adata_healthy.var['mt'] = all_adata_healthy.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(all_adata_healthy, qc_vars=['mt'], inplace=True)\n",
    "\n",
    "print(all_adata_healthy)\n",
    "display(all_adata_healthy.obs.head(3))\n",
    "display(all_adata_healthy.var.head(3))\n",
    "assert all_adata_healthy.obs_names.is_unique, \"ERROR: all_adata_healthy.obs_names are not unique!\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd54dc1",
   "metadata": {},
   "source": [
    "### Merge All togeteher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c9b2d0f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30771"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(all_adata_diseased.var_names) & set(all_adata_healthy.var_names)) #* 100 / len(set(all_adata_diseased.var_names))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "591a5687",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gc\n",
    "gc.collect()  # Run once\n",
    "gc.collect()  # Run TWICE for circular references\n",
    "gc.collect()  # Even three times for stubborn cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa793155",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 210416 × 30771\n",
      "    obs: 'pct_intronic', 'Neighborhood_name', 'Neighborhood_bootstrapping_probability', 'Class_name', 'Class_bootstrapping_probability', 'Subclass_name', 'Subclass_bootstrapping_probability', 'Group_name', 'Group_bootstrapping_probability', 'Cluster_label', 'Cluster_name', 'Cluster_alias', 'Cluster_bootstrapping_probability', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'library', 'zone', 'donor_id', 'state'\n",
      "    var: 'feature_types', 'genome', 'gene_symbol'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:770: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    ['MIR1302-2HG', 'ENSG00000238009', 'ENSG00000239906', 'ENSG00000241860', 'ENSG00000286448']\n",
      "\n",
      "    Inferred to be: categorical\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pct_intronic</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>Class_name</th>\n",
       "      <th>Class_bootstrapping_probability</th>\n",
       "      <th>Subclass_name</th>\n",
       "      <th>Subclass_bootstrapping_probability</th>\n",
       "      <th>Group_name</th>\n",
       "      <th>Group_bootstrapping_probability</th>\n",
       "      <th>Cluster_label</th>\n",
       "      <th>...</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>log1p_total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>library</th>\n",
       "      <th>zone</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAACCAGCACTAGTCG-1-SI-TT-H1-sample_01-diseased</th>\n",
       "      <td>0.647217</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>STR D1 MSN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>STRv D1 MSN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLUST_0524</td>\n",
       "      <td>...</td>\n",
       "      <td>9.069583</td>\n",
       "      <td>52026.0</td>\n",
       "      <td>10.859518</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>0.019221</td>\n",
       "      <td>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1</td>\n",
       "      <td>3</td>\n",
       "      <td>sample_01</td>\n",
       "      <td>diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAACCCCAGCCATAAT-1-SI-TT-H1-sample_01-diseased</th>\n",
       "      <td>0.596793</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>STR D2 MSN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Tafa1 D2 Matrix MSN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLUST_0546</td>\n",
       "      <td>...</td>\n",
       "      <td>9.047704</td>\n",
       "      <td>43110.0</td>\n",
       "      <td>10.671533</td>\n",
       "      <td>57.0</td>\n",
       "      <td>4.060443</td>\n",
       "      <td>0.132220</td>\n",
       "      <td>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1</td>\n",
       "      <td>5</td>\n",
       "      <td>sample_01</td>\n",
       "      <td>diseased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAACTAGAGCCTAGAC-1-SI-TT-H1-sample_01-diseased</th>\n",
       "      <td>0.657474</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>STR Hybrid MSN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLUST_0515</td>\n",
       "      <td>...</td>\n",
       "      <td>8.819813</td>\n",
       "      <td>28455.0</td>\n",
       "      <td>10.256114</td>\n",
       "      <td>24.0</td>\n",
       "      <td>3.218876</td>\n",
       "      <td>0.084344</td>\n",
       "      <td>240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1</td>\n",
       "      <td>3</td>\n",
       "      <td>sample_01</td>\n",
       "      <td>diseased</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    pct_intronic  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      0.647217   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      0.596793   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      0.657474   \n",
       "\n",
       "                                                   Neighborhood_name  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   Subpallium GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   Subpallium GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...   Subpallium GABA   \n",
       "\n",
       "                                                    Neighborhood_bootstrapping_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                     1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                     1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                     1.0   \n",
       "\n",
       "                                                     Class_name  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CN LGE GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CN LGE GABA   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CN LGE GABA   \n",
       "\n",
       "                                                    Class_bootstrapping_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "\n",
       "                                                     Subclass_name  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      STR D1 MSN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      STR D2 MSN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  STR Hybrid MSN   \n",
       "\n",
       "                                                    Subclass_bootstrapping_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                 1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                 1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                                 1.0   \n",
       "\n",
       "                                                             Group_name  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...          STRv D1 MSN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  Tafa1 D2 Matrix MSN   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...    STRv D1 NUDAP MSN   \n",
       "\n",
       "                                                    Group_bootstrapping_probability  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                              1.0   \n",
       "\n",
       "                                                            Cluster_label  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CS20250428_CLUST_0524   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CS20250428_CLUST_0546   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  CS20250428_CLUST_0515   \n",
       "\n",
       "                                                    ...  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  ...   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  ...   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  ...   \n",
       "\n",
       "                                                   log1p_n_genes_by_counts  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                9.069583   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                9.047704   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...                8.819813   \n",
       "\n",
       "                                                   total_counts  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      52026.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      43110.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...      28455.0   \n",
       "\n",
       "                                                    log1p_total_counts  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...           10.859518   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...           10.671533   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...           10.256114   \n",
       "\n",
       "                                                    total_counts_mt  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...             10.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...             57.0   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...             24.0   \n",
       "\n",
       "                                                    log1p_total_counts_mt  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...               2.397895   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...               4.060443   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...               3.218876   \n",
       "\n",
       "                                                    pct_counts_mt  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...       0.019221   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...       0.132220   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...       0.084344   \n",
       "\n",
       "                                                                                     library  \\\n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1   \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...  240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1   \n",
       "\n",
       "                                                    zone   donor_id     state  \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...     3  sample_01  diseased  \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...     5  sample_01  diseased  \n",
       "240805_SL-EXD_0328_B22FKKYLT4---SI-TT-H1---AAAC...     3  sample_01  diseased  \n",
       "\n",
       "[3 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239906</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000239906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   feature_types  genome      gene_symbol\n",
       "MIR1302-2HG      Gene Expression  GRCh38      MIR1302-2HG\n",
       "ENSG00000238009  Gene Expression  GRCh38  ENSG00000238009\n",
       "ENSG00000239906  Gene Expression  GRCh38  ENSG00000239906"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "important_columns = [\"donor_id\", \"library\", \"log1p_n_genes_by_counts\", \"pct_counts_mt\", \"zone\"]\n",
    "\n",
    "\n",
    "# Save var info from first dataset before concat\n",
    "# So all dataset will have same metadata for genes\n",
    "var_info = all_adata_diseased.var[['feature_types', 'genome', 'gene_symbol']].copy()\n",
    "\n",
    "# Concate\n",
    "adata = sc.concat([all_adata_diseased, all_adata_healthy], \n",
    "                  axis=0, # concatenate along observations, i.e., stacking cells)\n",
    "                  index_unique='-' , # causes new barcodes to be orig_idx + '-' + key\n",
    "                  keys=[\"diseased\", \"healthy\"],\n",
    "                  label=\"state\",\n",
    "                  join='inner')  # keeps COMMON genes\n",
    "#del all_adata_diseased, all_adata_healthy\n",
    "\n",
    "# Restore var info for common genes\n",
    "adata.var = var_info.loc[adata.var_names]\n",
    "\n",
    "print(adata)\n",
    "display(adata.obs.head(3))\n",
    "display(adata.var.head(3))\n",
    "assert adata.obs_names.is_unique, \"ERROR: adata.obs_names are not unique!\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c36f5fe",
   "metadata": {},
   "source": [
    "### Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3ff57ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['zone'] = adata.obs['zone'].astype(str).astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c8ee3fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(f\"{BASE_PATH}/adata_for_DEG/adata_for_DEG_slidetag.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21096478",
   "metadata": {},
   "source": [
    "----\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "999e0f3e",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfe254c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1241, 3)\n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_csv(\"/home/gdallagl/myworkdir/XDP/utils/dorsal_ventral_gradient/exp_by_spline_combined.csv\")\n",
    "# df = df[(np.abs(df.mean_spearman) >= 0.1) & (df.combined_p <= 0.05)].sort_values(\"mean_spearman\")\n",
    "# print(df.shape)\n",
    "# df.to_csv(\"/home/gdallagl/myworkdir/XDP/utils/dorsal_ventral_gradient/exp_by_spline_combined_corr>0.1.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39f4fe49",
   "metadata": {},
   "source": [
    "# Convert to .qs to .h5ad"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xdp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
