{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "234ec8a2-5062-413d-9fec-44bb6c0d6bb1",
   "metadata": {},
   "source": [
    "# Goal\n",
    "This notebook will show how to map data using the MapMyCells data products released with the HMBA consensus Basal Ganglia taxonomy.\n",
    "\n",
    "[Original Notebook here](https://github.com/AllenInstitute/HMBA_BasalGanglia_Consensus_Taxonomy/blob/main/examples/using_MapMyCells_data.ipynb)\n",
    "\n",
    "# Install MapMyCell dependencies\n",
    "We are going to start by installing the `cell_type_mapper` (actual backend for MapMyCells) and `abc_atlas_access` (convenient download of other data provided by the AllenInstitute) libraries.\n",
    "```bash\n",
    "uv add \"cell_type_mapper@git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2\"\n",
    "uv add \"abc_atlas_access[notebooks] @ git+https://github.com/alleninstitute/abc_atlas_access.git\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9241df83",
   "metadata": {},
   "source": [
    "# Dowload mapping files for basal ganglia\n",
    "\n",
    "To run MapMyCells, you need two supporting data files: \n",
    "- [an HDF5 file](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/input_data_files/precomputed_stats_file.md) which defines your taxonomy and the average gene expression profiles of the taxons there in\n",
    "- [lookup table of marker genes](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/input_data_files/marker_gene_lookup.md) for your taxonomy. \n",
    "\n",
    "Let's download those two files for the **human** basal ganglia data. \n",
    "Refer back [to the main page](https://github.com/AllenInstitute/HMBA_BasalGanglia_Consensus_Taxonomy/blob/main/index.md#cell-type-mapping-with-mapmycells) for the locations of the relevant files for all three species (human, marmoset, and macaque).\n",
    "\n",
    "```bash\n",
    "wget https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.precomputed_stats.20250507.h5\n",
    "wget https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.query_markers.20250507.json\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "245dfb88",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e262f008-c7fa-4de3-bd90-8bed9d9f3ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from pathlib import Path\n",
    "\n",
    "import cell_type_mapper\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache\n",
    "from cell_type_mapper.cli.from_specified_markers import FromSpecifiedMarkersRunner\n",
    "\n",
    "# Hyperparamyters\n",
    "LIBRARY = \"H3\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0892870",
   "metadata": {},
   "source": [
    "# Prepare Query data\n",
    "\n",
    "Your query data must be:\n",
    "- **Format**: `.h5ad` file (AnnData format)\n",
    "- **Structure**: \n",
    "  - `X` layer with **RAW** gene expression data (cells × genes)\n",
    "  - `obs` with cell metadata\n",
    "  - `var` with gene names, **Ensembl IDs** for MapMyCells-supported taxonomies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a70bb8e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY_PATH=f\"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-{LIBRARY}/adata/raw_adata.h5ad\"\n",
    "\n",
    "# paths to files where mapping output will be written\n",
    "json_dst_path = str(Path(QUERY_PATH).parent.parent / \"map_my_cell\" / \"mapping.json\")\n",
    "csv_dst_path = str(Path(QUERY_PATH).parent.parent / \"map_my_cell\" / \"mapping.csv\")\n",
    "\n",
    "# updated adata path\n",
    "adata_labelled_path = str(Path(QUERY_PATH).parent / \"labelled_adata.h5ad\")\n",
    "\n",
    "# FIXED (always mappign againf basal ganglia)\n",
    "# the lookup table of marker genes which we downloaded from S3\n",
    "query_marker_path = \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.query_markers.20250507.json\"\n",
    "# the human-specific precomputed stats file which we downloaded from S3\n",
    "precomputed_path = \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.precomputed_stats.20250507.h5\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d116d9",
   "metadata": {},
   "source": [
    "# Run Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eacd0613-a698-4c7f-8224-ba15856c009b",
   "metadata": {},
   "source": [
    "Now we will actually [perform the mapping](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/mapping_cells.md)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "529c1e69-49e5-4c1a-aa25-6de2db06f413",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    # output paths\n",
    "    \"query_path\": QUERY_PATH,\n",
    "    \"extended_result_path\": json_dst_path,\n",
    "    \"csv_result_path\": csv_dst_path,\n",
    "    \"verbose_csv\": True,\n",
    "\n",
    "    # inout paths\n",
    "    \"query_markers\": {\n",
    "       \"serialized_lookup\": query_marker_path\n",
    "    },\n",
    "    \"precomputed_stats\": {\n",
    "        \"path\": precomputed_path\n",
    "    },\n",
    "\n",
    "    \"type_assignment\": {\n",
    "        \"n_processors\": 8,\n",
    "        \"normalization\": \"raw\", #Use raw counts (not normalized)\n",
    "        \"bootstrap_factor\": 0.5,\n",
    "        \"bootstrap_iteration\": 100\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d680bfe0-78fe-45a4-8815-5e229fcc4275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Running Hierarchical Mapping 1.5.2 with config ===\n",
      "{\n",
      "  \"query_path\": \"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H3/adata/raw_adata.h5ad\",\n",
      "  \"map_to_ensembl\": false,\n",
      "  \"obsm_clobber\": false,\n",
      "  \"verbose_csv\": true,\n",
      "  \"hdf5_result_path\": null,\n",
      "  \"max_gb\": 100.0,\n",
      "  \"query_markers\": {\n",
      "    \"serialized_lookup\": \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.query_markers.20250507.json\",\n",
      "    \"collapse_markers\": false,\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"log_level\": \"ERROR\",\n",
      "  \"verbose_stdout\": true,\n",
      "  \"obsm_key\": null,\n",
      "  \"summary_metadata_path\": null,\n",
      "  \"csv_result_path\": \"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H3/map_my_cell/mapping.csv\",\n",
      "  \"precomputed_stats\": {\n",
      "    \"path\": \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.precomputed_stats.20250507.h5\",\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"drop_level\": null,\n",
      "  \"cloud_safe\": false,\n",
      "  \"extended_result_path\": \"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H3/map_my_cell/mapping.json\",\n",
      "  \"nodes_to_drop\": null,\n",
      "  \"extended_result_dir\": null,\n",
      "  \"type_assignment\": {\n",
      "    \"bootstrap_iteration\": 100,\n",
      "    \"bootstrap_factor_lookup\": null,\n",
      "    \"min_markers\": 10,\n",
      "    \"n_processors\": 8,\n",
      "    \"chunk_size\": 10000,\n",
      "    \"normalization\": \"raw\",\n",
      "    \"bootstrap_factor\": 0.5,\n",
      "    \"n_runners_up\": 5,\n",
      "    \"rng_seed\": 11235813,\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"query_gene_id_col\": null,\n",
      "  \"tmp_dir\": null,\n",
      "  \"flatten\": false,\n",
      "  \"log_path\": null\n",
      "}\n",
      "ENV: is_torch_available: False\n",
      "ENV: is_cuda_available: False\n",
      "ENV: use_torch: False\n",
      "ENV: multiprocessing start method: fork\n",
      "ENV: Python version: 3.11.8 (main, Sep 24 2025, 14:13:47) [GCC 12.2.0]\n",
      "ENV: anndata version: 0.12.6\n",
      "ENV: numpy version: 2.2.6\n",
      "BENCHMARK: spent 5.1451e-04 seconds validating config and copying data\n",
      "using ../Human.precomputed_stats.20250507.h5 for precomputed_stats\n",
      "reading taxonomy_tree from ../Human.precomputed_stats.20250507.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: numpy's internal parallelization is enabled. This could cause independent worker processes to compete for resources, degrading performance. We recommend setting the following environment variables to '1' to improve performance\n",
      "{\n",
      "  \"NUMEXPR_NUM_THREADS\": \"\",\n",
      "  \"MKL_NUM_THREADS\": \"\",\n",
      "  \"OMP_NUM_THREADS\": \"\"\n",
      "}\n",
      "  warnings.warn(msg)\n",
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/cli/cli_log.py:104: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  self.env(f\"anndata version: {anndata.__version__}\")\n",
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 32 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BENCHMARK: spent 1.4915e-01 seconds creating query marker cache\n",
      "Running CPU implementation of type assignment.\n",
      "BENCHMARK: spent 5.3183e+02 seconds assigning cell types\n",
      "Writing marker genes to output file\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "runner = FromSpecifiedMarkersRunner(\n",
    "    args=[],\n",
    "    input_data=config\n",
    ")\n",
    "runner.run()\n",
    "print(\"Done!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153381c4-5655-4d41-9b09-b54289427e86",
   "metadata": {},
   "source": [
    "# Output of mapping file\n",
    "\n",
    "The results of our mapping are now in two files: the csv file pointed to by `csv_dst_path` and the JSON file pointed to by `json_dst_path`. Dedicated documentation of the the contents of the mapping output [can be found here.](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/output.md)\n",
    "\n",
    "## CSV output file\n",
    "\n",
    "The CSV file is effectively just a dataframe. For every cell at every taxonomy level, you have its assigned cell type (both as a guaranteed unique \"label\" and a more human readable \"name\") along with quality metrics assessing the confidence in the mapping (see the detailed documentation above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75fde950-89eb-46e6-b9ce-84b8f98bd281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Neighborhood_label</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>Neighborhood_aggregate_probability</th>\n",
       "      <th>Neighborhood_correlation_coefficient</th>\n",
       "      <th>Class_label</th>\n",
       "      <th>Class_name</th>\n",
       "      <th>Class_bootstrapping_probability</th>\n",
       "      <th>Class_aggregate_probability</th>\n",
       "      <th>Class_correlation_coefficient</th>\n",
       "      <th>...</th>\n",
       "      <th>Group_name</th>\n",
       "      <th>Group_bootstrapping_probability</th>\n",
       "      <th>Group_aggregate_probability</th>\n",
       "      <th>Group_correlation_coefficient</th>\n",
       "      <th>Cluster_label</th>\n",
       "      <th>Cluster_name</th>\n",
       "      <th>Cluster_alias</th>\n",
       "      <th>Cluster_bootstrapping_probability</th>\n",
       "      <th>Cluster_aggregate_probability</th>\n",
       "      <th>Cluster_correlation_coefficient</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGAACAGAC-1</th>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6523</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6431</td>\n",
       "      <td>...</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5333</td>\n",
       "      <td>CS20250428_CLUST_0516</td>\n",
       "      <td>Human-484</td>\n",
       "      <td>Human-484</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.4764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGAATATCC-1</th>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7357</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6947</td>\n",
       "      <td>...</td>\n",
       "      <td>OT D1 ICj</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.6176</td>\n",
       "      <td>CS20250428_CLUST_0338</td>\n",
       "      <td>Human-164</td>\n",
       "      <td>Human-164</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGAGTCGTC-1</th>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7695</td>\n",
       "      <td>CS20250428_CLASS_0003</td>\n",
       "      <td>CN LGE GABA</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7420</td>\n",
       "      <td>...</td>\n",
       "      <td>STRd D2 Striosome MSN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5643</td>\n",
       "      <td>CS20250428_CLUST_0485</td>\n",
       "      <td>Human-189</td>\n",
       "      <td>Human-189</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>0.4365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGATCCCAA-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7086</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>Astro-Epen</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7801</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9900</td>\n",
       "      <td>0.3414</td>\n",
       "      <td>CS20250428_CLUST_0255</td>\n",
       "      <td>Human-232</td>\n",
       "      <td>Human-232</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.9405</td>\n",
       "      <td>0.4905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGCAAGCGA-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5931</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>Astro-Epen</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6480</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.9100</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>CS20250428_CLUST_0249</td>\n",
       "      <td>Human-14</td>\n",
       "      <td>Human-14</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.4641</td>\n",
       "      <td>0.3844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTACGTGAT-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6906</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8336</td>\n",
       "      <td>...</td>\n",
       "      <td>Microglia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.6258</td>\n",
       "      <td>CS20250428_CLUST_0225</td>\n",
       "      <td>Human-543</td>\n",
       "      <td>Human-543</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.7100</td>\n",
       "      <td>0.3815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTAGGCATT-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6890</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8381</td>\n",
       "      <td>...</td>\n",
       "      <td>OPC</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.8231</td>\n",
       "      <td>CS20250428_CLUST_0232</td>\n",
       "      <td>Human-11</td>\n",
       "      <td>Human-11</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>0.3882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTATCGTGG-1</th>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>CS20250428_CLASS_0006</td>\n",
       "      <td>F M GABA</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.5724</td>\n",
       "      <td>...</td>\n",
       "      <td>ZI-HTH GABA</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.5238</td>\n",
       "      <td>0.4203</td>\n",
       "      <td>CS20250428_CLUST_0356</td>\n",
       "      <td>Human-521_1</td>\n",
       "      <td>Human-521_1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5238</td>\n",
       "      <td>0.4203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCATACCC-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7080</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8174</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.6100</td>\n",
       "      <td>0.5653</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6100</td>\n",
       "      <td>0.6234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCCTGCGC-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6961</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8671</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.7046</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36649 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Neighborhood_label Neighborhood_name  \\\n",
       "cell_id                                                       \n",
       "AAACCAAAGAACAGAC-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "AAACCAAAGAATATCC-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "AAACCAAAGAGTCGTC-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "AAACCAAAGATCCCAA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGCAAGCGA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "...                                   ...               ...   \n",
       "GTTGTGGGTACGTGAT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTAGGCATT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTATCGTGG-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "GTTGTGGGTCATACCC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTCCTGCGC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "\n",
       "                    Neighborhood_bootstrapping_probability  \\\n",
       "cell_id                                                      \n",
       "AAACCAAAGAACAGAC-1                                     1.0   \n",
       "AAACCAAAGAATATCC-1                                     1.0   \n",
       "AAACCAAAGAGTCGTC-1                                     1.0   \n",
       "AAACCAAAGATCCCAA-1                                     1.0   \n",
       "AAACCAAAGCAAGCGA-1                                     1.0   \n",
       "...                                                    ...   \n",
       "GTTGTGGGTACGTGAT-1                                     1.0   \n",
       "GTTGTGGGTAGGCATT-1                                     1.0   \n",
       "GTTGTGGGTATCGTGG-1                                     1.0   \n",
       "GTTGTGGGTCATACCC-1                                     1.0   \n",
       "GTTGTGGGTCCTGCGC-1                                     1.0   \n",
       "\n",
       "                    Neighborhood_aggregate_probability  \\\n",
       "cell_id                                                  \n",
       "AAACCAAAGAACAGAC-1                                 1.0   \n",
       "AAACCAAAGAATATCC-1                                 1.0   \n",
       "AAACCAAAGAGTCGTC-1                                 1.0   \n",
       "AAACCAAAGATCCCAA-1                                 1.0   \n",
       "AAACCAAAGCAAGCGA-1                                 1.0   \n",
       "...                                                ...   \n",
       "GTTGTGGGTACGTGAT-1                                 1.0   \n",
       "GTTGTGGGTAGGCATT-1                                 1.0   \n",
       "GTTGTGGGTATCGTGG-1                                 1.0   \n",
       "GTTGTGGGTCATACCC-1                                 1.0   \n",
       "GTTGTGGGTCCTGCGC-1                                 1.0   \n",
       "\n",
       "                    Neighborhood_correlation_coefficient  \\\n",
       "cell_id                                                    \n",
       "AAACCAAAGAACAGAC-1                                0.6523   \n",
       "AAACCAAAGAATATCC-1                                0.7357   \n",
       "AAACCAAAGAGTCGTC-1                                0.7695   \n",
       "AAACCAAAGATCCCAA-1                                0.7086   \n",
       "AAACCAAAGCAAGCGA-1                                0.5931   \n",
       "...                                                  ...   \n",
       "GTTGTGGGTACGTGAT-1                                0.6906   \n",
       "GTTGTGGGTAGGCATT-1                                0.6890   \n",
       "GTTGTGGGTATCGTGG-1                                0.6163   \n",
       "GTTGTGGGTCATACCC-1                                0.7080   \n",
       "GTTGTGGGTCCTGCGC-1                                0.6961   \n",
       "\n",
       "                              Class_label   Class_name  \\\n",
       "cell_id                                                  \n",
       "AAACCAAAGAACAGAC-1  CS20250428_CLASS_0003  CN LGE GABA   \n",
       "AAACCAAAGAATATCC-1  CS20250428_CLASS_0003  CN LGE GABA   \n",
       "AAACCAAAGAGTCGTC-1  CS20250428_CLASS_0003  CN LGE GABA   \n",
       "AAACCAAAGATCCCAA-1  CS20250428_CLASS_0000   Astro-Epen   \n",
       "AAACCAAAGCAAGCGA-1  CS20250428_CLASS_0000   Astro-Epen   \n",
       "...                                   ...          ...   \n",
       "GTTGTGGGTACGTGAT-1  CS20250428_CLASS_0008       Immune   \n",
       "GTTGTGGGTAGGCATT-1  CS20250428_CLASS_0010    OPC-Oligo   \n",
       "GTTGTGGGTATCGTGG-1  CS20250428_CLASS_0006     F M GABA   \n",
       "GTTGTGGGTCATACCC-1  CS20250428_CLASS_0010    OPC-Oligo   \n",
       "GTTGTGGGTCCTGCGC-1  CS20250428_CLASS_0010    OPC-Oligo   \n",
       "\n",
       "                    Class_bootstrapping_probability  \\\n",
       "cell_id                                               \n",
       "AAACCAAAGAACAGAC-1                             1.00   \n",
       "AAACCAAAGAATATCC-1                             1.00   \n",
       "AAACCAAAGAGTCGTC-1                             1.00   \n",
       "AAACCAAAGATCCCAA-1                             1.00   \n",
       "AAACCAAAGCAAGCGA-1                             1.00   \n",
       "...                                             ...   \n",
       "GTTGTGGGTACGTGAT-1                             1.00   \n",
       "GTTGTGGGTAGGCATT-1                             1.00   \n",
       "GTTGTGGGTATCGTGG-1                             0.54   \n",
       "GTTGTGGGTCATACCC-1                             1.00   \n",
       "GTTGTGGGTCCTGCGC-1                             1.00   \n",
       "\n",
       "                    Class_aggregate_probability  \\\n",
       "cell_id                                           \n",
       "AAACCAAAGAACAGAC-1                         1.00   \n",
       "AAACCAAAGAATATCC-1                         1.00   \n",
       "AAACCAAAGAGTCGTC-1                         1.00   \n",
       "AAACCAAAGATCCCAA-1                         1.00   \n",
       "AAACCAAAGCAAGCGA-1                         1.00   \n",
       "...                                         ...   \n",
       "GTTGTGGGTACGTGAT-1                         1.00   \n",
       "GTTGTGGGTAGGCATT-1                         1.00   \n",
       "GTTGTGGGTATCGTGG-1                         0.54   \n",
       "GTTGTGGGTCATACCC-1                         1.00   \n",
       "GTTGTGGGTCCTGCGC-1                         1.00   \n",
       "\n",
       "                    Class_correlation_coefficient  ...             Group_name  \\\n",
       "cell_id                                            ...                          \n",
       "AAACCAAAGAACAGAC-1                         0.6431  ...      STRv D1 NUDAP MSN   \n",
       "AAACCAAAGAATATCC-1                         0.6947  ...              OT D1 ICj   \n",
       "AAACCAAAGAGTCGTC-1                         0.7420  ...  STRd D2 Striosome MSN   \n",
       "AAACCAAAGATCCCAA-1                         0.7801  ...              Astrocyte   \n",
       "AAACCAAAGCAAGCGA-1                         0.6480  ...              Astrocyte   \n",
       "...                                           ...  ...                    ...   \n",
       "GTTGTGGGTACGTGAT-1                         0.8336  ...              Microglia   \n",
       "GTTGTGGGTAGGCATT-1                         0.8381  ...                    OPC   \n",
       "GTTGTGGGTATCGTGG-1                         0.5724  ...            ZI-HTH GABA   \n",
       "GTTGTGGGTCATACCC-1                         0.8174  ...           Oligo OPALIN   \n",
       "GTTGTGGGTCCTGCGC-1                         0.8671  ...           Oligo OPALIN   \n",
       "\n",
       "                   Group_bootstrapping_probability  \\\n",
       "cell_id                                              \n",
       "AAACCAAAGAACAGAC-1                            1.00   \n",
       "AAACCAAAGAATATCC-1                            1.00   \n",
       "AAACCAAAGAGTCGTC-1                            1.00   \n",
       "AAACCAAAGATCCCAA-1                            0.99   \n",
       "AAACCAAAGCAAGCGA-1                            0.91   \n",
       "...                                            ...   \n",
       "GTTGTGGGTACGTGAT-1                            1.00   \n",
       "GTTGTGGGTAGGCATT-1                            1.00   \n",
       "GTTGTGGGTATCGTGG-1                            0.97   \n",
       "GTTGTGGGTCATACCC-1                            0.61   \n",
       "GTTGTGGGTCCTGCGC-1                            1.00   \n",
       "\n",
       "                    Group_aggregate_probability  \\\n",
       "cell_id                                           \n",
       "AAACCAAAGAACAGAC-1                       1.0000   \n",
       "AAACCAAAGAATATCC-1                       1.0000   \n",
       "AAACCAAAGAGTCGTC-1                       1.0000   \n",
       "AAACCAAAGATCCCAA-1                       0.9900   \n",
       "AAACCAAAGCAAGCGA-1                       0.9100   \n",
       "...                                         ...   \n",
       "GTTGTGGGTACGTGAT-1                       1.0000   \n",
       "GTTGTGGGTAGGCATT-1                       1.0000   \n",
       "GTTGTGGGTATCGTGG-1                       0.5238   \n",
       "GTTGTGGGTCATACCC-1                       0.6100   \n",
       "GTTGTGGGTCCTGCGC-1                       1.0000   \n",
       "\n",
       "                    Group_correlation_coefficient          Cluster_label  \\\n",
       "cell_id                                                                    \n",
       "AAACCAAAGAACAGAC-1                         0.5333  CS20250428_CLUST_0516   \n",
       "AAACCAAAGAATATCC-1                         0.6176  CS20250428_CLUST_0338   \n",
       "AAACCAAAGAGTCGTC-1                         0.5643  CS20250428_CLUST_0485   \n",
       "AAACCAAAGATCCCAA-1                         0.3414  CS20250428_CLUST_0255   \n",
       "AAACCAAAGCAAGCGA-1                         0.2416  CS20250428_CLUST_0249   \n",
       "...                                           ...                    ...   \n",
       "GTTGTGGGTACGTGAT-1                         0.6258  CS20250428_CLUST_0225   \n",
       "GTTGTGGGTAGGCATT-1                         0.8231  CS20250428_CLUST_0232   \n",
       "GTTGTGGGTATCGTGG-1                         0.4203  CS20250428_CLUST_0356   \n",
       "GTTGTGGGTCATACCC-1                         0.5653  CS20250428_CLUST_0227   \n",
       "GTTGTGGGTCCTGCGC-1                         0.7046  CS20250428_CLUST_0227   \n",
       "\n",
       "                   Cluster_name Cluster_alias  \\\n",
       "cell_id                                         \n",
       "AAACCAAAGAACAGAC-1    Human-484     Human-484   \n",
       "AAACCAAAGAATATCC-1    Human-164     Human-164   \n",
       "AAACCAAAGAGTCGTC-1    Human-189     Human-189   \n",
       "AAACCAAAGATCCCAA-1    Human-232     Human-232   \n",
       "AAACCAAAGCAAGCGA-1     Human-14      Human-14   \n",
       "...                         ...           ...   \n",
       "GTTGTGGGTACGTGAT-1    Human-543     Human-543   \n",
       "GTTGTGGGTAGGCATT-1     Human-11      Human-11   \n",
       "GTTGTGGGTATCGTGG-1  Human-521_1   Human-521_1   \n",
       "GTTGTGGGTCATACCC-1      Human-1       Human-1   \n",
       "GTTGTGGGTCCTGCGC-1      Human-1       Human-1   \n",
       "\n",
       "                    Cluster_bootstrapping_probability  \\\n",
       "cell_id                                                 \n",
       "AAACCAAAGAACAGAC-1                               1.00   \n",
       "AAACCAAAGAATATCC-1                               1.00   \n",
       "AAACCAAAGAGTCGTC-1                               0.47   \n",
       "AAACCAAAGATCCCAA-1                               0.95   \n",
       "AAACCAAAGCAAGCGA-1                               0.51   \n",
       "...                                               ...   \n",
       "GTTGTGGGTACGTGAT-1                               0.71   \n",
       "GTTGTGGGTAGGCATT-1                               0.87   \n",
       "GTTGTGGGTATCGTGG-1                               1.00   \n",
       "GTTGTGGGTCATACCC-1                               1.00   \n",
       "GTTGTGGGTCCTGCGC-1                               1.00   \n",
       "\n",
       "                    Cluster_aggregate_probability  \\\n",
       "cell_id                                             \n",
       "AAACCAAAGAACAGAC-1                         1.0000   \n",
       "AAACCAAAGAATATCC-1                         1.0000   \n",
       "AAACCAAAGAGTCGTC-1                         0.4700   \n",
       "AAACCAAAGATCCCAA-1                         0.9405   \n",
       "AAACCAAAGCAAGCGA-1                         0.4641   \n",
       "...                                           ...   \n",
       "GTTGTGGGTACGTGAT-1                         0.7100   \n",
       "GTTGTGGGTAGGCATT-1                         0.8700   \n",
       "GTTGTGGGTATCGTGG-1                         0.5238   \n",
       "GTTGTGGGTCATACCC-1                         0.6100   \n",
       "GTTGTGGGTCCTGCGC-1                         1.0000   \n",
       "\n",
       "                    Cluster_correlation_coefficient  \n",
       "cell_id                                              \n",
       "AAACCAAAGAACAGAC-1                           0.4764  \n",
       "AAACCAAAGAATATCC-1                           0.5817  \n",
       "AAACCAAAGAGTCGTC-1                           0.4365  \n",
       "AAACCAAAGATCCCAA-1                           0.4905  \n",
       "AAACCAAAGCAAGCGA-1                           0.3844  \n",
       "...                                             ...  \n",
       "GTTGTGGGTACGTGAT-1                           0.3815  \n",
       "GTTGTGGGTAGGCATT-1                           0.3882  \n",
       "GTTGTGGGTATCGTGG-1                           0.4203  \n",
       "GTTGTGGGTCATACCC-1                           0.6234  \n",
       "GTTGTGGGTCCTGCGC-1                           0.5201  \n",
       "\n",
       "[36649 rows x 26 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mapping_csv = pd.read_csv(csv_dst_path, comment='#')\n",
    "mapping_csv = mapping_csv.set_index(\"cell_id\")\n",
    "\n",
    "display(mapping_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77e978f",
   "metadata": {},
   "source": [
    "# Add Metadata to adata and Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53b1263e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>pct_intronic</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>dbscan_clusters</th>\n",
       "      <th>dbscan_score</th>\n",
       "      <th>has_spatial</th>\n",
       "      <th>Neighborhood_label</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>...</th>\n",
       "      <th>Group_name</th>\n",
       "      <th>Group_bootstrapping_probability</th>\n",
       "      <th>Group_aggregate_probability</th>\n",
       "      <th>Group_correlation_coefficient</th>\n",
       "      <th>Cluster_label</th>\n",
       "      <th>Cluster_name</th>\n",
       "      <th>Cluster_alias</th>\n",
       "      <th>Cluster_bootstrapping_probability</th>\n",
       "      <th>Cluster_aggregate_probability</th>\n",
       "      <th>Cluster_correlation_coefficient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGAACAGAC-1</th>\n",
       "      <td>36750.336875</td>\n",
       "      <td>-4918.947027</td>\n",
       "      <td>0.582893</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.914737</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>STRv D1 NUDAP MSN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5333</td>\n",
       "      <td>CS20250428_CLUST_0516</td>\n",
       "      <td>Human-484</td>\n",
       "      <td>Human-484</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.4764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGAATATCC-1</th>\n",
       "      <td>6546.427751</td>\n",
       "      <td>-8964.535170</td>\n",
       "      <td>0.676476</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.949614</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>OT D1 ICj</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.6176</td>\n",
       "      <td>CS20250428_CLUST_0338</td>\n",
       "      <td>Human-164</td>\n",
       "      <td>Human-164</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGAGTCGTC-1</th>\n",
       "      <td>45734.792830</td>\n",
       "      <td>8427.068976</td>\n",
       "      <td>0.712612</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.978896</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>STRd D2 Striosome MSN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5643</td>\n",
       "      <td>CS20250428_CLUST_0485</td>\n",
       "      <td>Human-189</td>\n",
       "      <td>Human-189</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>0.4365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGATCCCAA-1</th>\n",
       "      <td>51201.227539</td>\n",
       "      <td>525.607101</td>\n",
       "      <td>0.686855</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.960045</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.9900</td>\n",
       "      <td>0.3414</td>\n",
       "      <td>CS20250428_CLUST_0255</td>\n",
       "      <td>Human-232</td>\n",
       "      <td>Human-232</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.9405</td>\n",
       "      <td>0.4905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGCAAGCGA-1</th>\n",
       "      <td>8939.060270</td>\n",
       "      <td>5043.015355</td>\n",
       "      <td>0.672605</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.973961</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.9100</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>CS20250428_CLUST_0249</td>\n",
       "      <td>Human-14</td>\n",
       "      <td>Human-14</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.4641</td>\n",
       "      <td>0.3844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTACGTGAT-1</th>\n",
       "      <td>7312.232149</td>\n",
       "      <td>5689.570899</td>\n",
       "      <td>0.671489</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.790263</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Microglia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.6258</td>\n",
       "      <td>CS20250428_CLUST_0225</td>\n",
       "      <td>Human-543</td>\n",
       "      <td>Human-543</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.7100</td>\n",
       "      <td>0.3815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTAGGCATT-1</th>\n",
       "      <td>11313.567114</td>\n",
       "      <td>8499.633020</td>\n",
       "      <td>0.751973</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.966612</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>OPC</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.8231</td>\n",
       "      <td>CS20250428_CLUST_0232</td>\n",
       "      <td>Human-11</td>\n",
       "      <td>Human-11</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.8700</td>\n",
       "      <td>0.3882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTATCGTGG-1</th>\n",
       "      <td>31393.199291</td>\n",
       "      <td>-120.827487</td>\n",
       "      <td>0.610678</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.963507</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0002</td>\n",
       "      <td>Subpallium GABA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>ZI-HTH GABA</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.5238</td>\n",
       "      <td>0.4203</td>\n",
       "      <td>CS20250428_CLUST_0356</td>\n",
       "      <td>Human-521_1</td>\n",
       "      <td>Human-521_1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5238</td>\n",
       "      <td>0.4203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCATACCC-1</th>\n",
       "      <td>46188.098628</td>\n",
       "      <td>-7745.600240</td>\n",
       "      <td>0.691346</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.903554</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.6100</td>\n",
       "      <td>0.5653</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6100</td>\n",
       "      <td>0.6234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCCTGCGC-1</th>\n",
       "      <td>43884.375265</td>\n",
       "      <td>14036.552216</td>\n",
       "      <td>0.646723</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>0.655188</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.7046</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.5201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36649 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               x             y  pct_intronic  is_cell  \\\n",
       "AAACCAAAGAACAGAC-1  36750.336875  -4918.947027      0.582893     True   \n",
       "AAACCAAAGAATATCC-1   6546.427751  -8964.535170      0.676476     True   \n",
       "AAACCAAAGAGTCGTC-1  45734.792830   8427.068976      0.712612     True   \n",
       "AAACCAAAGATCCCAA-1  51201.227539    525.607101      0.686855     True   \n",
       "AAACCAAAGCAAGCGA-1   8939.060270   5043.015355      0.672605     True   \n",
       "...                          ...           ...           ...      ...   \n",
       "GTTGTGGGTACGTGAT-1   7312.232149   5689.570899      0.671489     True   \n",
       "GTTGTGGGTAGGCATT-1  11313.567114   8499.633020      0.751973     True   \n",
       "GTTGTGGGTATCGTGG-1  31393.199291   -120.827487      0.610678     True   \n",
       "GTTGTGGGTCATACCC-1  46188.098628  -7745.600240      0.691346     True   \n",
       "GTTGTGGGTCCTGCGC-1  43884.375265  14036.552216      0.646723     True   \n",
       "\n",
       "                    dbscan_clusters  dbscan_score  has_spatial  \\\n",
       "AAACCAAAGAACAGAC-1                1      0.914737         True   \n",
       "AAACCAAAGAATATCC-1                1      0.949614         True   \n",
       "AAACCAAAGAGTCGTC-1                1      0.978896         True   \n",
       "AAACCAAAGATCCCAA-1                1      0.960045         True   \n",
       "AAACCAAAGCAAGCGA-1                1      0.973961         True   \n",
       "...                             ...           ...          ...   \n",
       "GTTGTGGGTACGTGAT-1                1      0.790263         True   \n",
       "GTTGTGGGTAGGCATT-1                1      0.966612         True   \n",
       "GTTGTGGGTATCGTGG-1                1      0.963507         True   \n",
       "GTTGTGGGTCATACCC-1                1      0.903554         True   \n",
       "GTTGTGGGTCCTGCGC-1                1      0.655188         True   \n",
       "\n",
       "                       Neighborhood_label Neighborhood_name  \\\n",
       "AAACCAAAGAACAGAC-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "AAACCAAAGAATATCC-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "AAACCAAAGAGTCGTC-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "AAACCAAAGATCCCAA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGCAAGCGA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "...                                   ...               ...   \n",
       "GTTGTGGGTACGTGAT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTAGGCATT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTATCGTGG-1  CS20250428_NEIGH_0002   Subpallium GABA   \n",
       "GTTGTGGGTCATACCC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTCCTGCGC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "\n",
       "                    Neighborhood_bootstrapping_probability  ...  \\\n",
       "AAACCAAAGAACAGAC-1                                     1.0  ...   \n",
       "AAACCAAAGAATATCC-1                                     1.0  ...   \n",
       "AAACCAAAGAGTCGTC-1                                     1.0  ...   \n",
       "AAACCAAAGATCCCAA-1                                     1.0  ...   \n",
       "AAACCAAAGCAAGCGA-1                                     1.0  ...   \n",
       "...                                                    ...  ...   \n",
       "GTTGTGGGTACGTGAT-1                                     1.0  ...   \n",
       "GTTGTGGGTAGGCATT-1                                     1.0  ...   \n",
       "GTTGTGGGTATCGTGG-1                                     1.0  ...   \n",
       "GTTGTGGGTCATACCC-1                                     1.0  ...   \n",
       "GTTGTGGGTCCTGCGC-1                                     1.0  ...   \n",
       "\n",
       "                               Group_name  Group_bootstrapping_probability  \\\n",
       "AAACCAAAGAACAGAC-1      STRv D1 NUDAP MSN                             1.00   \n",
       "AAACCAAAGAATATCC-1              OT D1 ICj                             1.00   \n",
       "AAACCAAAGAGTCGTC-1  STRd D2 Striosome MSN                             1.00   \n",
       "AAACCAAAGATCCCAA-1              Astrocyte                             0.99   \n",
       "AAACCAAAGCAAGCGA-1              Astrocyte                             0.91   \n",
       "...                                   ...                              ...   \n",
       "GTTGTGGGTACGTGAT-1              Microglia                             1.00   \n",
       "GTTGTGGGTAGGCATT-1                    OPC                             1.00   \n",
       "GTTGTGGGTATCGTGG-1            ZI-HTH GABA                             0.97   \n",
       "GTTGTGGGTCATACCC-1           Oligo OPALIN                             0.61   \n",
       "GTTGTGGGTCCTGCGC-1           Oligo OPALIN                             1.00   \n",
       "\n",
       "                   Group_aggregate_probability Group_correlation_coefficient  \\\n",
       "AAACCAAAGAACAGAC-1                      1.0000                        0.5333   \n",
       "AAACCAAAGAATATCC-1                      1.0000                        0.6176   \n",
       "AAACCAAAGAGTCGTC-1                      1.0000                        0.5643   \n",
       "AAACCAAAGATCCCAA-1                      0.9900                        0.3414   \n",
       "AAACCAAAGCAAGCGA-1                      0.9100                        0.2416   \n",
       "...                                        ...                           ...   \n",
       "GTTGTGGGTACGTGAT-1                      1.0000                        0.6258   \n",
       "GTTGTGGGTAGGCATT-1                      1.0000                        0.8231   \n",
       "GTTGTGGGTATCGTGG-1                      0.5238                        0.4203   \n",
       "GTTGTGGGTCATACCC-1                      0.6100                        0.5653   \n",
       "GTTGTGGGTCCTGCGC-1                      1.0000                        0.7046   \n",
       "\n",
       "                            Cluster_label  Cluster_name  Cluster_alias  \\\n",
       "AAACCAAAGAACAGAC-1  CS20250428_CLUST_0516     Human-484      Human-484   \n",
       "AAACCAAAGAATATCC-1  CS20250428_CLUST_0338     Human-164      Human-164   \n",
       "AAACCAAAGAGTCGTC-1  CS20250428_CLUST_0485     Human-189      Human-189   \n",
       "AAACCAAAGATCCCAA-1  CS20250428_CLUST_0255     Human-232      Human-232   \n",
       "AAACCAAAGCAAGCGA-1  CS20250428_CLUST_0249      Human-14       Human-14   \n",
       "...                                   ...           ...            ...   \n",
       "GTTGTGGGTACGTGAT-1  CS20250428_CLUST_0225     Human-543      Human-543   \n",
       "GTTGTGGGTAGGCATT-1  CS20250428_CLUST_0232      Human-11       Human-11   \n",
       "GTTGTGGGTATCGTGG-1  CS20250428_CLUST_0356   Human-521_1    Human-521_1   \n",
       "GTTGTGGGTCATACCC-1  CS20250428_CLUST_0227       Human-1        Human-1   \n",
       "GTTGTGGGTCCTGCGC-1  CS20250428_CLUST_0227       Human-1        Human-1   \n",
       "\n",
       "                   Cluster_bootstrapping_probability  \\\n",
       "AAACCAAAGAACAGAC-1                              1.00   \n",
       "AAACCAAAGAATATCC-1                              1.00   \n",
       "AAACCAAAGAGTCGTC-1                              0.47   \n",
       "AAACCAAAGATCCCAA-1                              0.95   \n",
       "AAACCAAAGCAAGCGA-1                              0.51   \n",
       "...                                              ...   \n",
       "GTTGTGGGTACGTGAT-1                              0.71   \n",
       "GTTGTGGGTAGGCATT-1                              0.87   \n",
       "GTTGTGGGTATCGTGG-1                              1.00   \n",
       "GTTGTGGGTCATACCC-1                              1.00   \n",
       "GTTGTGGGTCCTGCGC-1                              1.00   \n",
       "\n",
       "                   Cluster_aggregate_probability  \\\n",
       "AAACCAAAGAACAGAC-1                        1.0000   \n",
       "AAACCAAAGAATATCC-1                        1.0000   \n",
       "AAACCAAAGAGTCGTC-1                        0.4700   \n",
       "AAACCAAAGATCCCAA-1                        0.9405   \n",
       "AAACCAAAGCAAGCGA-1                        0.4641   \n",
       "...                                          ...   \n",
       "GTTGTGGGTACGTGAT-1                        0.7100   \n",
       "GTTGTGGGTAGGCATT-1                        0.8700   \n",
       "GTTGTGGGTATCGTGG-1                        0.5238   \n",
       "GTTGTGGGTCATACCC-1                        0.6100   \n",
       "GTTGTGGGTCCTGCGC-1                        1.0000   \n",
       "\n",
       "                    Cluster_correlation_coefficient  \n",
       "AAACCAAAGAACAGAC-1                           0.4764  \n",
       "AAACCAAAGAATATCC-1                           0.5817  \n",
       "AAACCAAAGAGTCGTC-1                           0.4365  \n",
       "AAACCAAAGATCCCAA-1                           0.4905  \n",
       "AAACCAAAGCAAGCGA-1                           0.3844  \n",
       "...                                             ...  \n",
       "GTTGTGGGTACGTGAT-1                           0.3815  \n",
       "GTTGTGGGTAGGCATT-1                           0.3882  \n",
       "GTTGTGGGTATCGTGG-1                           0.4203  \n",
       "GTTGTGGGTCATACCC-1                           0.6234  \n",
       "GTTGTGGGTCCTGCGC-1                           0.5201  \n",
       "\n",
       "[36649 rows x 33 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original cells: 36649\n",
      "Mapped cells: 36649\n"
     ]
    }
   ],
   "source": [
    "# Read adata\n",
    "query_adata = sc.read_h5ad(QUERY_PATH)\n",
    "\n",
    "# Merge metadata\n",
    "query_adata.obs = query_adata.obs.join(mapping_csv, how=\"left\")\n",
    "display(query_adata.obs)\n",
    "\n",
    "# save \n",
    "query_adata.write(adata_labelled_path)\n",
    "\n",
    "\n",
    "n_cells_original = len(query_adata.obs)\n",
    "n_cells_mapped = mapping_csv.index.nunique()\n",
    "print(f\"Original cells: {n_cells_original}\")\n",
    "print(f\"Mapped cells: {n_cells_mapped}\")\n",
    "assert n_cells_original == n_cells_mapped, \"Mapping incomplete!\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a043a1",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0357fa7b-be1a-4906-b47d-f93031cc2be9",
   "metadata": {},
   "source": [
    "## Optional: JSON output file\n",
    "\n",
    "The JSON output contains everything in the CSV file, along with helpful metadata about your mapping run as [documented here](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/output.md#json-output-file)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86762dd5-e1b8-4dff-b356-72a2e1fb2a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open(json_dst_path, 'rb') as src:\n",
    "#     mapping_json = json.load(src)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dce8801-d6c5-4173-9179-55bd2e0c10cd",
   "metadata": {},
   "source": [
    "For instance, to see the configuration parameters corresponding to your mapping run, you can look at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9dfe4251-9411-4c4d-9c38-5d4bd0dee286",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(json.dumps(mapping_json['config'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "417ab995-549b-4da5-b802-3a578c2e27d2",
   "metadata": {},
   "source": [
    "The actual cell type assignments are stored as a list under `'results'` as in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c56ac2cf-c34f-4d80-ae9b-2a1c7dcf5cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(json.dumps(mapping_json['results'][0], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa09ca1a-af74-4800-93c2-92dfdf0c646d",
   "metadata": {},
   "source": [
    "One complication is that the cell type assignments are only referred to by their unique machine-readable labels in this file. Fortunately, the cell type taxonomy, along with the mapping between machine- and human-readable cell type labels is also provided in this file. The `TaxonomyTree` class provides a helpful interface with that data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "12374288-0bbc-4fc7-ae96-1aa6e14b2eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from cell_type_mapper.taxonomy.taxonomy_tree import TaxonomyTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "85f6c1b9-6d8b-4c12-a3f5-075668b2afd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# taxonomy = TaxonomyTree(data=mapping_json['taxonomy_tree'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "beba2c6b-1945-44df-82ed-c4b85917b3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# taxonomy.label_to_name(level='CCN20250428_LEVEL_3', label='CS20250428_GROUP_0025')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xdp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
