{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "745fda9f",
   "metadata": {},
   "source": [
    "# DEG Analays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ef174cfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "# from pydeseq2.dds import DeseqDataSet\n",
    "# from pydeseq2.ds import DeseqStats\n",
    "# import decoupler as dc\n",
    "\n",
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "BASE_PATH = os.getenv(\"BASE_PATH_FOR_SAVING\")\n",
    "DISEASED_SAMPLES_FOLDER = f\"{BASE_PATH}/diseased\"\n",
    "MSN_BICAN_FOLDER = \"/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad\"\n",
    "\n",
    "# Cell type annotation to use for DEG\n",
    "CT_ANNOTATION_FOR_DEG = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab42d2a7",
   "metadata": {},
   "source": [
    "# Create Cell Type specific adata\n",
    "\n",
    "Create adata obj with both dieseased anf helathy sample but for a single cell type.\n",
    "\n",
    "Helthy and diseased samples are saved in differt locations and have differ nomenclture for columns.\n",
    "\n",
    "In addition, diseased are split btu sample, while healthy are already split by cell type but not by sample."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43e5f5fb",
   "metadata": {},
   "source": [
    "### Diseased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8f22d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_diseased_list = []\n",
    "sample_diseased_list = []\n",
    "\n",
    "# Load Diseased data (each adata is a differt sample/patient)\n",
    "for sample_folder in [p for p in Path(DISEASED_SAMPLES_FOLDER).iterdir() if p.is_dir()]:\n",
    "\n",
    "    adata_path = sample_folder / \"adata\" / \"all_lib_adata_zoned.h5ad\"\n",
    "\n",
    "    adata_tmp = sc.read_h5ad(adata_path, backed=\"r\")\n",
    "\n",
    "    # print(adata_tmp.obs.columns)\n",
    "    # display(adata_tmp.obs)\n",
    "    # print(sample_folder)\n",
    "\n",
    "    # restart from raw counts\n",
    "    adata_minimal = sc.AnnData(\n",
    "        X=adata_tmp.layers[\"counts\"].copy(),                   \n",
    "        obs=adata_tmp.obs.copy(),\n",
    "        var=adata_tmp.var.copy()\n",
    "    )\n",
    "    del adata_tmp\n",
    "\n",
    "    # add sample:\n",
    "    sample_diseased_list.append(sample_folder.name)\n",
    "\n",
    "    # add ot list\n",
    "    adata_diseased_list.append(adata_minimal)\n",
    "\n",
    "    #break\n",
    "\n",
    "adata_diseased_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c3698f",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'adata_diseased_list' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 3\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Save var info from first dataset before concat\u001b[39;00m\n\u001b[32m      2\u001b[39m \u001b[38;5;66;03m# So all dataset will have same metadata for genes\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m var_info = \u001b[43madata_diseased_list\u001b[49m[\u001b[32m0\u001b[39m].var[[\u001b[33m'\u001b[39m\u001b[33mfeature_types\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mgenome\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mgene_symbol\u001b[39m\u001b[33m'\u001b[39m]].copy()\n\u001b[32m      5\u001b[39m \u001b[38;5;66;03m# Concate\u001b[39;00m\n\u001b[32m      6\u001b[39m all_adata_diseased = sc.concat(adata_diseased_list, \n\u001b[32m      7\u001b[39m                   axis=\u001b[32m0\u001b[39m, \u001b[38;5;66;03m# concatenate along observations, i.e., stacking cells)\u001b[39;00m\n\u001b[32m      8\u001b[39m                   index_unique=\u001b[33m'\u001b[39m\u001b[33m-\u001b[39m\u001b[33m'\u001b[39m , \u001b[38;5;66;03m# causes new barcodes to be orig_idx + '-' + key\u001b[39;00m\n\u001b[32m      9\u001b[39m                   keys=sample_diseased_list,\n\u001b[32m     10\u001b[39m                   label=\u001b[33m\"\u001b[39m\u001b[33mdonor_id\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     11\u001b[39m                   join=\u001b[33m'\u001b[39m\u001b[33minner\u001b[39m\u001b[33m'\u001b[39m)  \u001b[38;5;66;03m# keeps COMMON genes\u001b[39;00m\n",
      "\u001b[31mNameError\u001b[39m: name 'adata_diseased_list' is not defined"
     ]
    }
   ],
   "source": [
    "# Save var info from first dataset before concat\n",
    "# So all dataset will have same metadata for genes\n",
    "var_info = adata_diseased_list[0].var[['feature_types', 'genome', 'gene_symbol']].copy()\n",
    "\n",
    "# Concate\n",
    "all_adata_diseased = sc.concat(adata_diseased_list, \n",
    "                  axis=0, # concatenate along observations, i.e., stacking cells)\n",
    "                  index_unique='-' , # causes new barcodes to be orig_idx + '-' + key\n",
    "                  keys=sample_diseased_list,\n",
    "                  label=\"donor_id\",\n",
    "                  join='inner')  # keeps COMMON genes\n",
    "del adata_diseased_list\n",
    "\n",
    "# Restore var info for common genes\n",
    "all_adata_diseased.var = var_info.loc[all_adata_diseased.var_names]\n",
    "\n",
    "print(all_adata_diseased)\n",
    "display(all_adata_diseased.obs.head(3))\n",
    "display(all_adata_diseased.var.head(3))\n",
    "assert all_adata_diseased.obs_names.is_unique, \"ERROR: all_adata_diseased.obs_names are not unique!\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44edae10",
   "metadata": {},
   "source": [
    "### Healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793f9bd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_healthy_list = []\n",
    "\n",
    "# Load Helahy data (each adata is a differt cell type)\n",
    "for adata_path in [p for p in Path(MSN_BICAN_FOLDER).iterdir() if not p.is_dir()]:\n",
    "\n",
    "    adata_tmp = sc.read_h5ad(adata_path)#, backed=\"r\")\n",
    "\n",
    "    # print(adata_tmp.obs.columns)\n",
    "    # display(adata_tmp.obs)\n",
    "    # print(sample_folder)\n",
    "\n",
    "    # restart from raw counts\n",
    "    adata_minimal = sc.AnnData(\n",
    "        X=adata_tmp.X.copy(),           # raw counts in X       \n",
    "        obs=adata_tmp.obs.copy(),\n",
    "        var=adata_tmp.var.copy()\n",
    "    )\n",
    "    del adata_tmp\n",
    "\n",
    "    # add ot list\n",
    "    adata_healthy_list.append(adata_minimal)\n",
    "\n",
    "adata_healthy_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "089b881c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Concate\n",
    "all_adata_healthy = sc.concat(adata_healthy_list, \n",
    "                  axis=0, # concatenate along observations, i.e., stacking cells)\n",
    "                  index_unique='-' , # causes new barcodes to be orig_idx + '-' + key\n",
    "                #   keys=sample_diseased_list,\n",
    "                #   label=\"donor_id\",\n",
    "                  join='inner')  # keeps COMMON genes\n",
    "del adata_healthy_list\n",
    "\n",
    "print(all_adata_healthy)\n",
    "display(all_adata_healthy.obs.head(3))\n",
    "display(all_adata_healthy.var.head(3))\n",
    "assert all_adata_healthy.obs_names.is_unique, \"ERROR: all_adata_healthy.obs_names are not unique!\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c3306a2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([301., 171., 130., 175., 325., 164., 383., 241., 143., 336., 260.,\n",
       "       231., 227., 131., 290., 277., 306., 253.,  61., 173., 313., 398.,\n",
       "       219., 307., 383., 285., 224., 211., 234., 198., 214., 241., 226.,\n",
       "       318., 300., 162., 321., 282., 179., 303., 424., 244., 329., 321.,\n",
       "       286., 123., 191., 213., 236., 317., 185., 200., 216., 245., 139.,\n",
       "        77., 203., 335., 144., 283., 209., 307., 225., 380., 379., 208.,\n",
       "       255., 199., 336.,  82., 201., 257., 162., 229., 299., 189., 211.,\n",
       "       362., 311.,  90., 208., 189., 352., 293., 381., 314., 153., 237.,\n",
       "       285., 408., 340., 287., 186., 292., 268., 242., 234., 199., 185.,\n",
       "       259.])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = sc.read_h5ad(\"/home/gdallagl/myworkdir/XDP/data/STR_zonation_references_h5ad/STRd_D1_Striosome_MSN.h5ad\", backed=\"r\")\n",
    "t.X[:100, :100].toarray().sum(axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd54dc1",
   "metadata": {},
   "source": [
    "### Merge All"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa793155",
   "metadata": {},
   "outputs": [],
   "source": [
    "important_columns = [\"donor_id\", \"library\", \"log1p_n_genes_by_counts\", \"log1p_total_counts_mt\", \"zone\"]\n",
    "\n",
    "\n",
    "# Save var info from first dataset before concat\n",
    "# So all dataset will have same metadata for genes\n",
    "var_info = all_adata_diseased.var[['feature_types', 'genome', 'gene_symbol']].copy()\n",
    "\n",
    "# Concate\n",
    "adata = sc.concat([all_adata_diseased, all_adata_healthy], \n",
    "                  axis=0, # concatenate along observations, i.e., stacking cells)\n",
    "                  index_unique='-' , # causes new barcodes to be orig_idx + '-' + key\n",
    "                  keys=[\"diseased\", \"healthy\"],\n",
    "                  label=\"state\",\n",
    "                  join='inner')  # keeps COMMON genes\n",
    "del adata_diseased_list\n",
    "\n",
    "# Restore var info for common genes\n",
    "adata.var = var_info.loc[adata.var_names]\n",
    "\n",
    "print(adata)\n",
    "display(adata.obs.head(3))\n",
    "display(adata.var.head(3))\n",
    "assert adata.obs_names.is_unique, \"ERROR: adata.obs_names are not unique!\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e72124e0",
   "metadata": {},
   "source": [
    "# Case I: healthy Vs Disease - All cells (GEX) - Desseq2 - ct X zone\n",
    "- Select specified (ct x zone) for each sample\n",
    "- Merge all selected cells in a single adata\n",
    "- Pseudobulk\n",
    "    - Sample level (ignore multiple lib â†’ just check visually)\n",
    "    - Sum raw counts\n",
    "    - 1 pseudo_cell per sample\n",
    "    - Min 10 cells to be a pseudocell\n",
    "    - Min 3 pseudocell per group\n",
    "- Run test with formula/cov\n",
    "    - design = ~ disease_state + log1p_n_genes_by_counts + pct_mito\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8efc300",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xdp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
