{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "234ec8a2-5062-413d-9fec-44bb6c0d6bb1",
   "metadata": {},
   "source": [
    "# Goal\n",
    "This notebook will show how to map data using the MapMyCells data products released with the HMBA consensus Basal Ganglia taxonomy.\n",
    "\n",
    "[Original Notebook here](https://github.com/AllenInstitute/HMBA_BasalGanglia_Consensus_Taxonomy/blob/main/examples/using_MapMyCells_data.ipynb)\n",
    "\n",
    "# Install MapMyCell dependencies\n",
    "We are going to start by installing the `cell_type_mapper` (actual backend for MapMyCells) and `abc_atlas_access` (convenient download of other data provided by the AllenInstitute) libraries.\n",
    "```bash\n",
    "uv add \"cell_type_mapper@git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2\"\n",
    "uv add \"abc_atlas_access[notebooks] @ git+https://github.com/alleninstitute/abc_atlas_access.git\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9241df83",
   "metadata": {},
   "source": [
    "# Dowload mapping files for basal ganglia\n",
    "\n",
    "To run MapMyCells, you need two supporting data files: \n",
    "- [an HDF5 file](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/input_data_files/precomputed_stats_file.md) which defines your taxonomy and the average gene expression profiles of the taxons there in\n",
    "- [lookup table of marker genes](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/input_data_files/marker_gene_lookup.md) for your taxonomy. \n",
    "\n",
    "Let's download those two files for the **human** basal ganglia data. \n",
    "Refer back [to the main page](https://github.com/AllenInstitute/HMBA_BasalGanglia_Consensus_Taxonomy/blob/main/index.md#cell-type-mapping-with-mapmycells) for the locations of the relevant files for all three species (human, marmoset, and macaque).\n",
    "\n",
    "```bash\n",
    "wget https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.precomputed_stats.20250507.h5\n",
    "wget https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.query_markers.20250507.json\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "245dfb88",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e262f008-c7fa-4de3-bd90-8bed9d9f3ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from pathlib import Path\n",
    "\n",
    "import cell_type_mapper\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache\n",
    "from cell_type_mapper.cli.from_specified_markers import FromSpecifiedMarkersRunner"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0892870",
   "metadata": {},
   "source": [
    "# Prepare Query data\n",
    "\n",
    "Your query data must be:\n",
    "- **Format**: `.h5ad` file (AnnData format)\n",
    "- **Structure**: \n",
    "  - `X` layer with **RAW** gene expression data (cells × genes)\n",
    "  - `obs` with cell metadata\n",
    "  - `var` with gene names, **Ensembl IDs** for MapMyCells-supported taxonomies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a70bb8e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY_PATH=\"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H1/adata/raw_adata.h5ad\"\n",
    "\n",
    "# paths to files where mapping output will be written\n",
    "json_dst_path = str(Path(QUERY_PATH).parent.parent / \"map_my_cell\" / \"mapping.json\")\n",
    "csv_dst_path = str(Path(QUERY_PATH).parent.parent / \"map_my_cell\" / \"mapping.csv\")\n",
    "\n",
    "# updated adata path\n",
    "adata_labelled_path = str(Path(QUERY_PATH).parent / \"labelled_adata.h5ad\")\n",
    "\n",
    "# FIXED (always mappign againf basal ganglia)\n",
    "# the lookup table of marker genes which we downloaded from S3\n",
    "query_marker_path = \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.query_markers.20250507.json\"\n",
    "# the human-specific precomputed stats file which we downloaded from S3\n",
    "precomputed_path = \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.precomputed_stats.20250507.h5\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d116d9",
   "metadata": {},
   "source": [
    "# Run Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eacd0613-a698-4c7f-8224-ba15856c009b",
   "metadata": {},
   "source": [
    "Now we will actually [perform the mapping](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/mapping_cells.md)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "529c1e69-49e5-4c1a-aa25-6de2db06f413",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    # output paths\n",
    "    \"query_path\": QUERY_PATH,\n",
    "    \"extended_result_path\": json_dst_path,\n",
    "    \"csv_result_path\": csv_dst_path,\n",
    "    \"verbose_csv\": True,\n",
    "\n",
    "    # inout paths\n",
    "    \"query_markers\": {\n",
    "       \"serialized_lookup\": query_marker_path\n",
    "    },\n",
    "    \"precomputed_stats\": {\n",
    "        \"path\": precomputed_path\n",
    "    },\n",
    "\n",
    "    \"type_assignment\": {\n",
    "        \"n_processors\": 4,\n",
    "        \"normalization\": \"raw\", #Use raw counts (not normalized)\n",
    "        \"bootstrap_factor\": 0.5,\n",
    "        \"bootstrap_iteration\": 100\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d680bfe0-78fe-45a4-8815-5e229fcc4275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Running Hierarchical Mapping 1.5.2 with config ===\n",
      "{\n",
      "  \"query_markers\": {\n",
      "    \"serialized_lookup\": \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.query_markers.20250507.json\",\n",
      "    \"collapse_markers\": false,\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"drop_level\": null,\n",
      "  \"query_path\": \"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H1/adata/raw_adata.h5ad\",\n",
      "  \"obsm_clobber\": false,\n",
      "  \"csv_result_path\": \"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H1/map_my_cell/mapping.csv\",\n",
      "  \"extended_result_dir\": null,\n",
      "  \"log_path\": null,\n",
      "  \"precomputed_stats\": {\n",
      "    \"log_level\": \"ERROR\",\n",
      "    \"path\": \"/home/gdallagl/myworkdir/XDP/data/AllenAtlas/BGT_human_20250507/Human.precomputed_stats.20250507.h5\"\n",
      "  },\n",
      "  \"hdf5_result_path\": null,\n",
      "  \"nodes_to_drop\": null,\n",
      "  \"flatten\": false,\n",
      "  \"verbose_csv\": true,\n",
      "  \"type_assignment\": {\n",
      "    \"n_processors\": 4,\n",
      "    \"chunk_size\": 10000,\n",
      "    \"bootstrap_factor\": 0.5,\n",
      "    \"min_markers\": 10,\n",
      "    \"bootstrap_factor_lookup\": null,\n",
      "    \"normalization\": \"raw\",\n",
      "    \"rng_seed\": 11235813,\n",
      "    \"bootstrap_iteration\": 100,\n",
      "    \"n_runners_up\": 5,\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"obsm_key\": null,\n",
      "  \"map_to_ensembl\": false,\n",
      "  \"cloud_safe\": false,\n",
      "  \"extended_result_path\": \"/home/gdallagl/myworkdir/XDP/data/XDP/disease/240805_SL-EXD_0328_B22FKKYLT4/SI-TT-H1/map_my_cell/mapping.json\",\n",
      "  \"tmp_dir\": null,\n",
      "  \"max_gb\": 100.0,\n",
      "  \"verbose_stdout\": true,\n",
      "  \"summary_metadata_path\": null,\n",
      "  \"log_level\": \"ERROR\",\n",
      "  \"query_gene_id_col\": null\n",
      "}\n",
      "ENV: is_torch_available: False\n",
      "ENV: is_cuda_available: False\n",
      "ENV: use_torch: False\n",
      "ENV: multiprocessing start method: fork\n",
      "ENV: Python version: 3.11.8 (main, Sep 24 2025, 14:13:47) [GCC 12.2.0]\n",
      "ENV: anndata version: 0.12.6\n",
      "ENV: numpy version: 2.2.6\n",
      "BENCHMARK: spent 4.0865e-04 seconds validating config and copying data\n",
      "using ../Human.precomputed_stats.20250507.h5 for precomputed_stats\n",
      "reading taxonomy_tree from ../Human.precomputed_stats.20250507.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: numpy's internal parallelization is enabled. This could cause independent worker processes to compete for resources, degrading performance. We recommend setting the following environment variables to '1' to improve performance\n",
      "{\n",
      "  \"NUMEXPR_NUM_THREADS\": \"\",\n",
      "  \"MKL_NUM_THREADS\": \"\",\n",
      "  \"OMP_NUM_THREADS\": \"\"\n",
      "}\n",
      "  warnings.warn(msg)\n",
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/cli/cli_log.py:104: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  self.env(f\"anndata version: {anndata.__version__}\")\n",
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "/home/gdallagl/myworkdir/XDP/.venv/lib/python3.11/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 32 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BENCHMARK: spent 1.4562e-01 seconds creating query marker cache\n",
      "Running CPU implementation of type assignment.\n",
      "BENCHMARK: spent 3.0735e+02 seconds assigning cell types\n",
      "Writing marker genes to output file\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "runner = FromSpecifiedMarkersRunner(\n",
    "    args=[],\n",
    "    input_data=config\n",
    ")\n",
    "runner.run()\n",
    "print(\"Done!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153381c4-5655-4d41-9b09-b54289427e86",
   "metadata": {},
   "source": [
    "# Output of mapping file\n",
    "\n",
    "The results of our mapping are now in two files: the csv file pointed to by `csv_dst_path` and the JSON file pointed to by `json_dst_path`. Dedicated documentation of the the contents of the mapping output [can be found here.](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/output.md)\n",
    "\n",
    "## CSV output file\n",
    "\n",
    "The CSV file is effectively just a dataframe. For every cell at every taxonomy level, you have its assigned cell type (both as a guaranteed unique \"label\" and a more human readable \"name\") along with quality metrics assessing the confidence in the mapping (see the detailed documentation above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75fde950-89eb-46e6-b9ce-84b8f98bd281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Neighborhood_label</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>Neighborhood_aggregate_probability</th>\n",
       "      <th>Neighborhood_correlation_coefficient</th>\n",
       "      <th>Class_label</th>\n",
       "      <th>Class_name</th>\n",
       "      <th>Class_bootstrapping_probability</th>\n",
       "      <th>Class_aggregate_probability</th>\n",
       "      <th>Class_correlation_coefficient</th>\n",
       "      <th>...</th>\n",
       "      <th>Group_name</th>\n",
       "      <th>Group_bootstrapping_probability</th>\n",
       "      <th>Group_aggregate_probability</th>\n",
       "      <th>Group_correlation_coefficient</th>\n",
       "      <th>Cluster_label</th>\n",
       "      <th>Cluster_name</th>\n",
       "      <th>Cluster_alias</th>\n",
       "      <th>Cluster_bootstrapping_probability</th>\n",
       "      <th>Cluster_aggregate_probability</th>\n",
       "      <th>Cluster_correlation_coefficient</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGATAACAG-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6303</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>Astro-Epen</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.6408</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.5962</td>\n",
       "      <td>CS20250428_CLUST_0253</td>\n",
       "      <td>Human-230</td>\n",
       "      <td>Human-230</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.5403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGCATGGAG-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7041</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>...</td>\n",
       "      <td>Microglia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5218</td>\n",
       "      <td>CS20250428_CLUST_0223</td>\n",
       "      <td>Human-532</td>\n",
       "      <td>Human-532</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.4024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGCGGAGT-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7610</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8914</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo PLEKHG1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7088</td>\n",
       "      <td>CS20250428_CLUST_0231</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.6570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGTCTATG-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>Astro-Epen</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6255</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5057</td>\n",
       "      <td>CS20250428_CLUST_0167</td>\n",
       "      <td>Human-159</td>\n",
       "      <td>Human-159</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.4633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGTTGTAT-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7130</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8124</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6399</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCCGTCCA-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7741</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8936</td>\n",
       "      <td>...</td>\n",
       "      <td>ImOligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>CS20250428_CLUST_0203</td>\n",
       "      <td>Human-53</td>\n",
       "      <td>Human-53</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.7143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCGAAGCC-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7561</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.9014</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.6007</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.8427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTGCACGTA-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7351</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8877</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7132</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTGGCGTGT-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5699</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>Astro-Epen</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6010</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.4413</td>\n",
       "      <td>CS20250428_CLUST_0168</td>\n",
       "      <td>Human-160</td>\n",
       "      <td>Human-160</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.4814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTGGTAATC-1</th>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6963</td>\n",
       "      <td>CS20250428_CLASS_0011</td>\n",
       "      <td>Vascular</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.8441</td>\n",
       "      <td>...</td>\n",
       "      <td>Endo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7683</td>\n",
       "      <td>CS20250428_CLUST_0184</td>\n",
       "      <td>Human-330</td>\n",
       "      <td>Human-330</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.5523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34704 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Neighborhood_label Neighborhood_name  \\\n",
       "cell_id                                                       \n",
       "AAACCAAAGATAACAG-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGCATGGAG-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGGCGGAGT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGGTCTATG-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGGTTGTAT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "...                                   ...               ...   \n",
       "GTTGTGGGTCCGTCCA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTCGAAGCC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTGCACGTA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTGGCGTGT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTGGTAATC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "\n",
       "                    Neighborhood_bootstrapping_probability  \\\n",
       "cell_id                                                      \n",
       "AAACCAAAGATAACAG-1                                     1.0   \n",
       "AAACCAAAGCATGGAG-1                                     1.0   \n",
       "AAACCAAAGGCGGAGT-1                                     1.0   \n",
       "AAACCAAAGGTCTATG-1                                     1.0   \n",
       "AAACCAAAGGTTGTAT-1                                     1.0   \n",
       "...                                                    ...   \n",
       "GTTGTGGGTCCGTCCA-1                                     1.0   \n",
       "GTTGTGGGTCGAAGCC-1                                     1.0   \n",
       "GTTGTGGGTGCACGTA-1                                     1.0   \n",
       "GTTGTGGGTGGCGTGT-1                                     1.0   \n",
       "GTTGTGGGTGGTAATC-1                                     1.0   \n",
       "\n",
       "                    Neighborhood_aggregate_probability  \\\n",
       "cell_id                                                  \n",
       "AAACCAAAGATAACAG-1                                 1.0   \n",
       "AAACCAAAGCATGGAG-1                                 1.0   \n",
       "AAACCAAAGGCGGAGT-1                                 1.0   \n",
       "AAACCAAAGGTCTATG-1                                 1.0   \n",
       "AAACCAAAGGTTGTAT-1                                 1.0   \n",
       "...                                                ...   \n",
       "GTTGTGGGTCCGTCCA-1                                 1.0   \n",
       "GTTGTGGGTCGAAGCC-1                                 1.0   \n",
       "GTTGTGGGTGCACGTA-1                                 1.0   \n",
       "GTTGTGGGTGGCGTGT-1                                 1.0   \n",
       "GTTGTGGGTGGTAATC-1                                 1.0   \n",
       "\n",
       "                    Neighborhood_correlation_coefficient  \\\n",
       "cell_id                                                    \n",
       "AAACCAAAGATAACAG-1                                0.6303   \n",
       "AAACCAAAGCATGGAG-1                                0.7041   \n",
       "AAACCAAAGGCGGAGT-1                                0.7610   \n",
       "AAACCAAAGGTCTATG-1                                0.6028   \n",
       "AAACCAAAGGTTGTAT-1                                0.7130   \n",
       "...                                                  ...   \n",
       "GTTGTGGGTCCGTCCA-1                                0.7741   \n",
       "GTTGTGGGTCGAAGCC-1                                0.7561   \n",
       "GTTGTGGGTGCACGTA-1                                0.7351   \n",
       "GTTGTGGGTGGCGTGT-1                                0.5699   \n",
       "GTTGTGGGTGGTAATC-1                                0.6963   \n",
       "\n",
       "                              Class_label  Class_name  \\\n",
       "cell_id                                                 \n",
       "AAACCAAAGATAACAG-1  CS20250428_CLASS_0000  Astro-Epen   \n",
       "AAACCAAAGCATGGAG-1  CS20250428_CLASS_0008      Immune   \n",
       "AAACCAAAGGCGGAGT-1  CS20250428_CLASS_0010   OPC-Oligo   \n",
       "AAACCAAAGGTCTATG-1  CS20250428_CLASS_0000  Astro-Epen   \n",
       "AAACCAAAGGTTGTAT-1  CS20250428_CLASS_0010   OPC-Oligo   \n",
       "...                                   ...         ...   \n",
       "GTTGTGGGTCCGTCCA-1  CS20250428_CLASS_0010   OPC-Oligo   \n",
       "GTTGTGGGTCGAAGCC-1  CS20250428_CLASS_0010   OPC-Oligo   \n",
       "GTTGTGGGTGCACGTA-1  CS20250428_CLASS_0010   OPC-Oligo   \n",
       "GTTGTGGGTGGCGTGT-1  CS20250428_CLASS_0000  Astro-Epen   \n",
       "GTTGTGGGTGGTAATC-1  CS20250428_CLASS_0011    Vascular   \n",
       "\n",
       "                    Class_bootstrapping_probability  \\\n",
       "cell_id                                               \n",
       "AAACCAAAGATAACAG-1                             0.66   \n",
       "AAACCAAAGCATGGAG-1                             1.00   \n",
       "AAACCAAAGGCGGAGT-1                             1.00   \n",
       "AAACCAAAGGTCTATG-1                             1.00   \n",
       "AAACCAAAGGTTGTAT-1                             1.00   \n",
       "...                                             ...   \n",
       "GTTGTGGGTCCGTCCA-1                             1.00   \n",
       "GTTGTGGGTCGAAGCC-1                             1.00   \n",
       "GTTGTGGGTGCACGTA-1                             1.00   \n",
       "GTTGTGGGTGGCGTGT-1                             1.00   \n",
       "GTTGTGGGTGGTAATC-1                             1.00   \n",
       "\n",
       "                    Class_aggregate_probability  \\\n",
       "cell_id                                           \n",
       "AAACCAAAGATAACAG-1                         0.66   \n",
       "AAACCAAAGCATGGAG-1                         1.00   \n",
       "AAACCAAAGGCGGAGT-1                         1.00   \n",
       "AAACCAAAGGTCTATG-1                         1.00   \n",
       "AAACCAAAGGTTGTAT-1                         1.00   \n",
       "...                                         ...   \n",
       "GTTGTGGGTCCGTCCA-1                         1.00   \n",
       "GTTGTGGGTCGAAGCC-1                         1.00   \n",
       "GTTGTGGGTGCACGTA-1                         1.00   \n",
       "GTTGTGGGTGGCGTGT-1                         1.00   \n",
       "GTTGTGGGTGGTAATC-1                         1.00   \n",
       "\n",
       "                    Class_correlation_coefficient  ...     Group_name  \\\n",
       "cell_id                                            ...                  \n",
       "AAACCAAAGATAACAG-1                         0.6408  ...      Astrocyte   \n",
       "AAACCAAAGCATGGAG-1                         0.8486  ...      Microglia   \n",
       "AAACCAAAGGCGGAGT-1                         0.8914  ...  Oligo PLEKHG1   \n",
       "AAACCAAAGGTCTATG-1                         0.6255  ...      Astrocyte   \n",
       "AAACCAAAGGTTGTAT-1                         0.8124  ...   Oligo OPALIN   \n",
       "...                                           ...  ...            ...   \n",
       "GTTGTGGGTCCGTCCA-1                         0.8936  ...        ImOligo   \n",
       "GTTGTGGGTCGAAGCC-1                         0.9014  ...   Oligo OPALIN   \n",
       "GTTGTGGGTGCACGTA-1                         0.8877  ...   Oligo OPALIN   \n",
       "GTTGTGGGTGGCGTGT-1                         0.6010  ...      Astrocyte   \n",
       "GTTGTGGGTGGTAATC-1                         0.8441  ...           Endo   \n",
       "\n",
       "                   Group_bootstrapping_probability  \\\n",
       "cell_id                                              \n",
       "AAACCAAAGATAACAG-1                            1.00   \n",
       "AAACCAAAGCATGGAG-1                            1.00   \n",
       "AAACCAAAGGCGGAGT-1                            1.00   \n",
       "AAACCAAAGGTCTATG-1                            1.00   \n",
       "AAACCAAAGGTTGTAT-1                            1.00   \n",
       "...                                            ...   \n",
       "GTTGTGGGTCCGTCCA-1                            1.00   \n",
       "GTTGTGGGTCGAAGCC-1                            0.98   \n",
       "GTTGTGGGTGCACGTA-1                            1.00   \n",
       "GTTGTGGGTGGCGTGT-1                            1.00   \n",
       "GTTGTGGGTGGTAATC-1                            1.00   \n",
       "\n",
       "                    Group_aggregate_probability  \\\n",
       "cell_id                                           \n",
       "AAACCAAAGATAACAG-1                         0.66   \n",
       "AAACCAAAGCATGGAG-1                         1.00   \n",
       "AAACCAAAGGCGGAGT-1                         1.00   \n",
       "AAACCAAAGGTCTATG-1                         1.00   \n",
       "AAACCAAAGGTTGTAT-1                         1.00   \n",
       "...                                         ...   \n",
       "GTTGTGGGTCCGTCCA-1                         1.00   \n",
       "GTTGTGGGTCGAAGCC-1                         0.98   \n",
       "GTTGTGGGTGCACGTA-1                         1.00   \n",
       "GTTGTGGGTGGCGTGT-1                         1.00   \n",
       "GTTGTGGGTGGTAATC-1                         1.00   \n",
       "\n",
       "                    Group_correlation_coefficient          Cluster_label  \\\n",
       "cell_id                                                                    \n",
       "AAACCAAAGATAACAG-1                         0.5962  CS20250428_CLUST_0253   \n",
       "AAACCAAAGCATGGAG-1                         0.5218  CS20250428_CLUST_0223   \n",
       "AAACCAAAGGCGGAGT-1                         0.7088  CS20250428_CLUST_0231   \n",
       "AAACCAAAGGTCTATG-1                         0.5057  CS20250428_CLUST_0167   \n",
       "AAACCAAAGGTTGTAT-1                         0.6399  CS20250428_CLUST_0227   \n",
       "...                                           ...                    ...   \n",
       "GTTGTGGGTCCGTCCA-1                         0.6727  CS20250428_CLUST_0203   \n",
       "GTTGTGGGTCGAAGCC-1                         0.6007  CS20250428_CLUST_0227   \n",
       "GTTGTGGGTGCACGTA-1                         0.7132  CS20250428_CLUST_0227   \n",
       "GTTGTGGGTGGCGTGT-1                         0.4413  CS20250428_CLUST_0168   \n",
       "GTTGTGGGTGGTAATC-1                         0.7683  CS20250428_CLUST_0184   \n",
       "\n",
       "                   Cluster_name Cluster_alias  \\\n",
       "cell_id                                         \n",
       "AAACCAAAGATAACAG-1    Human-230     Human-230   \n",
       "AAACCAAAGCATGGAG-1    Human-532     Human-532   \n",
       "AAACCAAAGGCGGAGT-1     Human-15      Human-15   \n",
       "AAACCAAAGGTCTATG-1    Human-159     Human-159   \n",
       "AAACCAAAGGTTGTAT-1      Human-1       Human-1   \n",
       "...                         ...           ...   \n",
       "GTTGTGGGTCCGTCCA-1     Human-53      Human-53   \n",
       "GTTGTGGGTCGAAGCC-1      Human-1       Human-1   \n",
       "GTTGTGGGTGCACGTA-1      Human-1       Human-1   \n",
       "GTTGTGGGTGGCGTGT-1    Human-160     Human-160   \n",
       "GTTGTGGGTGGTAATC-1    Human-330     Human-330   \n",
       "\n",
       "                    Cluster_bootstrapping_probability  \\\n",
       "cell_id                                                 \n",
       "AAACCAAAGATAACAG-1                               1.00   \n",
       "AAACCAAAGCATGGAG-1                               0.50   \n",
       "AAACCAAAGGCGGAGT-1                               0.99   \n",
       "AAACCAAAGGTCTATG-1                               0.98   \n",
       "AAACCAAAGGTTGTAT-1                               1.00   \n",
       "...                                               ...   \n",
       "GTTGTGGGTCCGTCCA-1                               0.80   \n",
       "GTTGTGGGTCGAAGCC-1                               1.00   \n",
       "GTTGTGGGTGCACGTA-1                               1.00   \n",
       "GTTGTGGGTGGCGTGT-1                               0.81   \n",
       "GTTGTGGGTGGTAATC-1                               0.86   \n",
       "\n",
       "                    Cluster_aggregate_probability  \\\n",
       "cell_id                                             \n",
       "AAACCAAAGATAACAG-1                           0.66   \n",
       "AAACCAAAGCATGGAG-1                           0.50   \n",
       "AAACCAAAGGCGGAGT-1                           0.99   \n",
       "AAACCAAAGGTCTATG-1                           0.98   \n",
       "AAACCAAAGGTTGTAT-1                           1.00   \n",
       "...                                           ...   \n",
       "GTTGTGGGTCCGTCCA-1                           0.80   \n",
       "GTTGTGGGTCGAAGCC-1                           0.98   \n",
       "GTTGTGGGTGCACGTA-1                           1.00   \n",
       "GTTGTGGGTGGCGTGT-1                           0.81   \n",
       "GTTGTGGGTGGTAATC-1                           0.86   \n",
       "\n",
       "                    Cluster_correlation_coefficient  \n",
       "cell_id                                              \n",
       "AAACCAAAGATAACAG-1                           0.5403  \n",
       "AAACCAAAGCATGGAG-1                           0.4024  \n",
       "AAACCAAAGGCGGAGT-1                           0.6570  \n",
       "AAACCAAAGGTCTATG-1                           0.4633  \n",
       "AAACCAAAGGTTGTAT-1                           0.5997  \n",
       "...                                             ...  \n",
       "GTTGTGGGTCCGTCCA-1                           0.7143  \n",
       "GTTGTGGGTCGAAGCC-1                           0.8427  \n",
       "GTTGTGGGTGCACGTA-1                           0.6591  \n",
       "GTTGTGGGTGGCGTGT-1                           0.4814  \n",
       "GTTGTGGGTGGTAATC-1                           0.5523  \n",
       "\n",
       "[34704 rows x 26 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mapping_csv = pd.read_csv(csv_dst_path, comment='#')\n",
    "mapping_csv = mapping_csv.set_index(\"cell_id\")\n",
    "\n",
    "display(mapping_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77e978f",
   "metadata": {},
   "source": [
    "# Add Metadata to adata and Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53b1263e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>pct_intronic</th>\n",
       "      <th>is_cell</th>\n",
       "      <th>Neighborhood_label</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>Neighborhood_aggregate_probability</th>\n",
       "      <th>Neighborhood_correlation_coefficient</th>\n",
       "      <th>Class_label</th>\n",
       "      <th>...</th>\n",
       "      <th>Group_name</th>\n",
       "      <th>Group_bootstrapping_probability</th>\n",
       "      <th>Group_aggregate_probability</th>\n",
       "      <th>Group_correlation_coefficient</th>\n",
       "      <th>Cluster_label</th>\n",
       "      <th>Cluster_name</th>\n",
       "      <th>Cluster_alias</th>\n",
       "      <th>Cluster_bootstrapping_probability</th>\n",
       "      <th>Cluster_aggregate_probability</th>\n",
       "      <th>Cluster_correlation_coefficient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGATAACAG-1</th>\n",
       "      <td>33013.428400</td>\n",
       "      <td>444.562211</td>\n",
       "      <td>0.585746</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6303</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.5962</td>\n",
       "      <td>CS20250428_CLUST_0253</td>\n",
       "      <td>Human-230</td>\n",
       "      <td>Human-230</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.5403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGCATGGAG-1</th>\n",
       "      <td>21225.159553</td>\n",
       "      <td>4193.340108</td>\n",
       "      <td>0.596834</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7041</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>...</td>\n",
       "      <td>Microglia</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5218</td>\n",
       "      <td>CS20250428_CLUST_0223</td>\n",
       "      <td>Human-532</td>\n",
       "      <td>Human-532</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.4024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGCGGAGT-1</th>\n",
       "      <td>16429.015488</td>\n",
       "      <td>3852.143369</td>\n",
       "      <td>0.716821</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7610</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo PLEKHG1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7088</td>\n",
       "      <td>CS20250428_CLUST_0231</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.6570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGTCTATG-1</th>\n",
       "      <td>43242.438818</td>\n",
       "      <td>-1822.457007</td>\n",
       "      <td>0.609493</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6028</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5057</td>\n",
       "      <td>CS20250428_CLUST_0167</td>\n",
       "      <td>Human-159</td>\n",
       "      <td>Human-159</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.4633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGTTGTAT-1</th>\n",
       "      <td>14118.369028</td>\n",
       "      <td>-4852.270550</td>\n",
       "      <td>0.588186</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7130</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6399</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.5997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCCGTCCA-1</th>\n",
       "      <td>57226.380126</td>\n",
       "      <td>-635.403689</td>\n",
       "      <td>0.597685</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7741</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>...</td>\n",
       "      <td>ImOligo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6727</td>\n",
       "      <td>CS20250428_CLUST_0203</td>\n",
       "      <td>Human-53</td>\n",
       "      <td>Human-53</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.7143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTCGAAGCC-1</th>\n",
       "      <td>46045.094685</td>\n",
       "      <td>3106.066837</td>\n",
       "      <td>0.630523</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7561</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.6007</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.8427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTGCACGTA-1</th>\n",
       "      <td>9851.267283</td>\n",
       "      <td>7168.714885</td>\n",
       "      <td>0.541554</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.7351</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>...</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7132</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.6591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTGGCGTGT-1</th>\n",
       "      <td>17765.143818</td>\n",
       "      <td>-4292.867209</td>\n",
       "      <td>0.630550</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5699</td>\n",
       "      <td>CS20250428_CLASS_0000</td>\n",
       "      <td>...</td>\n",
       "      <td>Astrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.4413</td>\n",
       "      <td>CS20250428_CLUST_0168</td>\n",
       "      <td>Human-160</td>\n",
       "      <td>Human-160</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.4814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTGTGGGTGGTAATC-1</th>\n",
       "      <td>42919.351395</td>\n",
       "      <td>1232.101028</td>\n",
       "      <td>0.351562</td>\n",
       "      <td>True</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6963</td>\n",
       "      <td>CS20250428_CLASS_0011</td>\n",
       "      <td>...</td>\n",
       "      <td>Endo</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.7683</td>\n",
       "      <td>CS20250428_CLUST_0184</td>\n",
       "      <td>Human-330</td>\n",
       "      <td>Human-330</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.5523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34704 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               x            y  pct_intronic  is_cell  \\\n",
       "AAACCAAAGATAACAG-1  33013.428400   444.562211      0.585746     True   \n",
       "AAACCAAAGCATGGAG-1  21225.159553  4193.340108      0.596834     True   \n",
       "AAACCAAAGGCGGAGT-1  16429.015488  3852.143369      0.716821     True   \n",
       "AAACCAAAGGTCTATG-1  43242.438818 -1822.457007      0.609493     True   \n",
       "AAACCAAAGGTTGTAT-1  14118.369028 -4852.270550      0.588186     True   \n",
       "...                          ...          ...           ...      ...   \n",
       "GTTGTGGGTCCGTCCA-1  57226.380126  -635.403689      0.597685     True   \n",
       "GTTGTGGGTCGAAGCC-1  46045.094685  3106.066837      0.630523     True   \n",
       "GTTGTGGGTGCACGTA-1   9851.267283  7168.714885      0.541554     True   \n",
       "GTTGTGGGTGGCGTGT-1  17765.143818 -4292.867209      0.630550     True   \n",
       "GTTGTGGGTGGTAATC-1  42919.351395  1232.101028      0.351562     True   \n",
       "\n",
       "                       Neighborhood_label Neighborhood_name  \\\n",
       "AAACCAAAGATAACAG-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGCATGGAG-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGGCGGAGT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGGTCTATG-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "AAACCAAAGGTTGTAT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "...                                   ...               ...   \n",
       "GTTGTGGGTCCGTCCA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTCGAAGCC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTGCACGTA-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTGGCGTGT-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "GTTGTGGGTGGTAATC-1  CS20250428_NEIGH_0001         Nonneuron   \n",
       "\n",
       "                    Neighborhood_bootstrapping_probability  \\\n",
       "AAACCAAAGATAACAG-1                                     1.0   \n",
       "AAACCAAAGCATGGAG-1                                     1.0   \n",
       "AAACCAAAGGCGGAGT-1                                     1.0   \n",
       "AAACCAAAGGTCTATG-1                                     1.0   \n",
       "AAACCAAAGGTTGTAT-1                                     1.0   \n",
       "...                                                    ...   \n",
       "GTTGTGGGTCCGTCCA-1                                     1.0   \n",
       "GTTGTGGGTCGAAGCC-1                                     1.0   \n",
       "GTTGTGGGTGCACGTA-1                                     1.0   \n",
       "GTTGTGGGTGGCGTGT-1                                     1.0   \n",
       "GTTGTGGGTGGTAATC-1                                     1.0   \n",
       "\n",
       "                    Neighborhood_aggregate_probability  \\\n",
       "AAACCAAAGATAACAG-1                                 1.0   \n",
       "AAACCAAAGCATGGAG-1                                 1.0   \n",
       "AAACCAAAGGCGGAGT-1                                 1.0   \n",
       "AAACCAAAGGTCTATG-1                                 1.0   \n",
       "AAACCAAAGGTTGTAT-1                                 1.0   \n",
       "...                                                ...   \n",
       "GTTGTGGGTCCGTCCA-1                                 1.0   \n",
       "GTTGTGGGTCGAAGCC-1                                 1.0   \n",
       "GTTGTGGGTGCACGTA-1                                 1.0   \n",
       "GTTGTGGGTGGCGTGT-1                                 1.0   \n",
       "GTTGTGGGTGGTAATC-1                                 1.0   \n",
       "\n",
       "                    Neighborhood_correlation_coefficient  \\\n",
       "AAACCAAAGATAACAG-1                                0.6303   \n",
       "AAACCAAAGCATGGAG-1                                0.7041   \n",
       "AAACCAAAGGCGGAGT-1                                0.7610   \n",
       "AAACCAAAGGTCTATG-1                                0.6028   \n",
       "AAACCAAAGGTTGTAT-1                                0.7130   \n",
       "...                                                  ...   \n",
       "GTTGTGGGTCCGTCCA-1                                0.7741   \n",
       "GTTGTGGGTCGAAGCC-1                                0.7561   \n",
       "GTTGTGGGTGCACGTA-1                                0.7351   \n",
       "GTTGTGGGTGGCGTGT-1                                0.5699   \n",
       "GTTGTGGGTGGTAATC-1                                0.6963   \n",
       "\n",
       "                              Class_label  ...     Group_name  \\\n",
       "AAACCAAAGATAACAG-1  CS20250428_CLASS_0000  ...      Astrocyte   \n",
       "AAACCAAAGCATGGAG-1  CS20250428_CLASS_0008  ...      Microglia   \n",
       "AAACCAAAGGCGGAGT-1  CS20250428_CLASS_0010  ...  Oligo PLEKHG1   \n",
       "AAACCAAAGGTCTATG-1  CS20250428_CLASS_0000  ...      Astrocyte   \n",
       "AAACCAAAGGTTGTAT-1  CS20250428_CLASS_0010  ...   Oligo OPALIN   \n",
       "...                                   ...  ...            ...   \n",
       "GTTGTGGGTCCGTCCA-1  CS20250428_CLASS_0010  ...        ImOligo   \n",
       "GTTGTGGGTCGAAGCC-1  CS20250428_CLASS_0010  ...   Oligo OPALIN   \n",
       "GTTGTGGGTGCACGTA-1  CS20250428_CLASS_0010  ...   Oligo OPALIN   \n",
       "GTTGTGGGTGGCGTGT-1  CS20250428_CLASS_0000  ...      Astrocyte   \n",
       "GTTGTGGGTGGTAATC-1  CS20250428_CLASS_0011  ...           Endo   \n",
       "\n",
       "                    Group_bootstrapping_probability  \\\n",
       "AAACCAAAGATAACAG-1                             1.00   \n",
       "AAACCAAAGCATGGAG-1                             1.00   \n",
       "AAACCAAAGGCGGAGT-1                             1.00   \n",
       "AAACCAAAGGTCTATG-1                             1.00   \n",
       "AAACCAAAGGTTGTAT-1                             1.00   \n",
       "...                                             ...   \n",
       "GTTGTGGGTCCGTCCA-1                             1.00   \n",
       "GTTGTGGGTCGAAGCC-1                             0.98   \n",
       "GTTGTGGGTGCACGTA-1                             1.00   \n",
       "GTTGTGGGTGGCGTGT-1                             1.00   \n",
       "GTTGTGGGTGGTAATC-1                             1.00   \n",
       "\n",
       "                    Group_aggregate_probability  \\\n",
       "AAACCAAAGATAACAG-1                         0.66   \n",
       "AAACCAAAGCATGGAG-1                         1.00   \n",
       "AAACCAAAGGCGGAGT-1                         1.00   \n",
       "AAACCAAAGGTCTATG-1                         1.00   \n",
       "AAACCAAAGGTTGTAT-1                         1.00   \n",
       "...                                         ...   \n",
       "GTTGTGGGTCCGTCCA-1                         1.00   \n",
       "GTTGTGGGTCGAAGCC-1                         0.98   \n",
       "GTTGTGGGTGCACGTA-1                         1.00   \n",
       "GTTGTGGGTGGCGTGT-1                         1.00   \n",
       "GTTGTGGGTGGTAATC-1                         1.00   \n",
       "\n",
       "                    Group_correlation_coefficient          Cluster_label  \\\n",
       "AAACCAAAGATAACAG-1                         0.5962  CS20250428_CLUST_0253   \n",
       "AAACCAAAGCATGGAG-1                         0.5218  CS20250428_CLUST_0223   \n",
       "AAACCAAAGGCGGAGT-1                         0.7088  CS20250428_CLUST_0231   \n",
       "AAACCAAAGGTCTATG-1                         0.5057  CS20250428_CLUST_0167   \n",
       "AAACCAAAGGTTGTAT-1                         0.6399  CS20250428_CLUST_0227   \n",
       "...                                           ...                    ...   \n",
       "GTTGTGGGTCCGTCCA-1                         0.6727  CS20250428_CLUST_0203   \n",
       "GTTGTGGGTCGAAGCC-1                         0.6007  CS20250428_CLUST_0227   \n",
       "GTTGTGGGTGCACGTA-1                         0.7132  CS20250428_CLUST_0227   \n",
       "GTTGTGGGTGGCGTGT-1                         0.4413  CS20250428_CLUST_0168   \n",
       "GTTGTGGGTGGTAATC-1                         0.7683  CS20250428_CLUST_0184   \n",
       "\n",
       "                   Cluster_name  Cluster_alias  \\\n",
       "AAACCAAAGATAACAG-1    Human-230      Human-230   \n",
       "AAACCAAAGCATGGAG-1    Human-532      Human-532   \n",
       "AAACCAAAGGCGGAGT-1     Human-15       Human-15   \n",
       "AAACCAAAGGTCTATG-1    Human-159      Human-159   \n",
       "AAACCAAAGGTTGTAT-1      Human-1        Human-1   \n",
       "...                         ...            ...   \n",
       "GTTGTGGGTCCGTCCA-1     Human-53       Human-53   \n",
       "GTTGTGGGTCGAAGCC-1      Human-1        Human-1   \n",
       "GTTGTGGGTGCACGTA-1      Human-1        Human-1   \n",
       "GTTGTGGGTGGCGTGT-1    Human-160      Human-160   \n",
       "GTTGTGGGTGGTAATC-1    Human-330      Human-330   \n",
       "\n",
       "                    Cluster_bootstrapping_probability  \\\n",
       "AAACCAAAGATAACAG-1                               1.00   \n",
       "AAACCAAAGCATGGAG-1                               0.50   \n",
       "AAACCAAAGGCGGAGT-1                               0.99   \n",
       "AAACCAAAGGTCTATG-1                               0.98   \n",
       "AAACCAAAGGTTGTAT-1                               1.00   \n",
       "...                                               ...   \n",
       "GTTGTGGGTCCGTCCA-1                               0.80   \n",
       "GTTGTGGGTCGAAGCC-1                               1.00   \n",
       "GTTGTGGGTGCACGTA-1                               1.00   \n",
       "GTTGTGGGTGGCGTGT-1                               0.81   \n",
       "GTTGTGGGTGGTAATC-1                               0.86   \n",
       "\n",
       "                    Cluster_aggregate_probability  \\\n",
       "AAACCAAAGATAACAG-1                           0.66   \n",
       "AAACCAAAGCATGGAG-1                           0.50   \n",
       "AAACCAAAGGCGGAGT-1                           0.99   \n",
       "AAACCAAAGGTCTATG-1                           0.98   \n",
       "AAACCAAAGGTTGTAT-1                           1.00   \n",
       "...                                           ...   \n",
       "GTTGTGGGTCCGTCCA-1                           0.80   \n",
       "GTTGTGGGTCGAAGCC-1                           0.98   \n",
       "GTTGTGGGTGCACGTA-1                           1.00   \n",
       "GTTGTGGGTGGCGTGT-1                           0.81   \n",
       "GTTGTGGGTGGTAATC-1                           0.86   \n",
       "\n",
       "                   Cluster_correlation_coefficient  \n",
       "AAACCAAAGATAACAG-1                          0.5403  \n",
       "AAACCAAAGCATGGAG-1                          0.4024  \n",
       "AAACCAAAGGCGGAGT-1                          0.6570  \n",
       "AAACCAAAGGTCTATG-1                          0.4633  \n",
       "AAACCAAAGGTTGTAT-1                          0.5997  \n",
       "...                                            ...  \n",
       "GTTGTGGGTCCGTCCA-1                          0.7143  \n",
       "GTTGTGGGTCGAAGCC-1                          0.8427  \n",
       "GTTGTGGGTGCACGTA-1                          0.6591  \n",
       "GTTGTGGGTGGCGTGT-1                          0.4814  \n",
       "GTTGTGGGTGGTAATC-1                          0.5523  \n",
       "\n",
       "[34704 rows x 30 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original cells: 34704\n",
      "Mapped cells: 34704\n"
     ]
    }
   ],
   "source": [
    "# Read adata\n",
    "query_adata = sc.read_h5ad(QUERY_PATH)\n",
    "\n",
    "# Merge metadata\n",
    "query_adata.obs = query_adata.obs.join(mapping_csv, how=\"left\")\n",
    "display(query_adata.obs)\n",
    "\n",
    "# save \n",
    "query_adata.write(adata_labelled_path)\n",
    "\n",
    "\n",
    "n_cells_original = len(query_adata.obs)\n",
    "n_cells_mapped = mapping_csv.index.nunique()\n",
    "print(f\"Original cells: {n_cells_original}\")\n",
    "print(f\"Mapped cells: {n_cells_mapped}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a043a1",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8e1262b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'stop' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mstop\u001b[49m\n",
      "\u001b[31mNameError\u001b[39m: name 'stop' is not defined"
     ]
    }
   ],
   "source": [
    "stop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0357fa7b-be1a-4906-b47d-f93031cc2be9",
   "metadata": {},
   "source": [
    "## Optional: JSON output file\n",
    "\n",
    "The JSON output contains everything in the CSV file, along with helpful metadata about your mapping run as [documented here](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/output.md#json-output-file)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86762dd5-e1b8-4dff-b356-72a2e1fb2a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(json_dst_path, 'rb') as src:\n",
    "    mapping_json = json.load(src)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dce8801-d6c5-4173-9179-55bd2e0c10cd",
   "metadata": {},
   "source": [
    "For instance, to see the configuration parameters corresponding to your mapping run, you can look at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dfe4251-9411-4c4d-9c38-5d4bd0dee286",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(mapping_json['config'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "417ab995-549b-4da5-b802-3a578c2e27d2",
   "metadata": {},
   "source": [
    "The actual cell type assignments are stored as a list under `'results'` as in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c56ac2cf-c34f-4d80-ae9b-2a1c7dcf5cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(mapping_json['results'][0], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa09ca1a-af74-4800-93c2-92dfdf0c646d",
   "metadata": {},
   "source": [
    "One complication is that the cell type assignments are only referred to by their unique machine-readable labels in this file. Fortunately, the cell type taxonomy, along with the mapping between machine- and human-readable cell type labels is also provided in this file. The `TaxonomyTree` class provides a helpful interface with that data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12374288-0bbc-4fc7-ae96-1aa6e14b2eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cell_type_mapper.taxonomy.taxonomy_tree import TaxonomyTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85f6c1b9-6d8b-4c12-a3f5-075668b2afd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy = TaxonomyTree(data=mapping_json['taxonomy_tree'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beba2c6b-1945-44df-82ed-c4b85917b3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy.label_to_name(level='CCN20250428_LEVEL_3', label='CS20250428_GROUP_0025')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xdp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
